<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Molecular Input/Output &mdash; CDPKit 1.3.0.dev1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/tabs.css?v=4c969af8" />
      <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=bd7b91d8" />

  
    <link rel="shortcut icon" href="../_static/favicon.svg"/>
    <link rel="canonical" href="https://cdpkit.org/master/cdpl_python_tutorial/cdpl_python_tutorial.html" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=41a6cf89"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../_static/tabs.js?v=3ee01567"></script>
        <script src="../_static/docs-versions-menu.js?v=bb46260f"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="CDPL Python Cookbook" href="../cdpl_python_cookbook/index.html" />
    <link rel="prev" title="CDPL Python Tutorial" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #aaaaaa" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../applications/index.html">Applications</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">CDPL Python Tutorial</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">1. Molecular Input/Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reading-molecules-from-an-input-file">1.1. Reading molecules from an input file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#retrieving-structure-data-from-mdl-sd-files">1.2. Retrieving structure data from MDL SD-files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#writing-molecules-to-an-output-file">1.3. Writing molecules to an output file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#description-of-structural-atom-environments-as-smiles-strings">1.4. Description of structural atom environments as SMILES strings</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chembl-molecule-standardization-and-parent-structure-extraction">1.5. ChEMBL molecule standardization and parent structure extraction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#standardization-prediction-of-protonation-states">1.6. Standardization/prediction of protonation states</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#d-structure-and-conformer-ensemble-generation">2. 3D Structure and Conformer Ensemble Generation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generating-single-low-energy-3d-structures">2.1. Generating single low-energy 3D structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generating-conformer-ensembles">2.2. Generating conformer ensembles</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#pharmacophore-generation-and-processing">3. Pharmacophore Generation and Processing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pharmacophore-generation-for-single-conformer-molecules">3.1. Pharmacophore generation for single-conformer molecules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pharmacophore-generation-for-multi-conformer-molecules">3.2. Pharmacophore generation for multi-conformer molecules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generating-ligand-receptor-interaction-pharmacophores">3.3. Generating ligand-receptor interaction pharmacophores</a></li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-of-molecules-to-a-reference-pharmacophore">3.4. Alignment of molecules to a reference pharmacophore</a></li>
<li class="toctree-l3"><a class="reference internal" href="#retrieving-information-about-pharmacophores-and-features">3.5. Retrieving information about pharmacophores and features</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculation-of-atom-properties">4. Calculation of Atom Properties</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calculation-of-atom-classification-properties">4.1. Calculation of atom classification properties</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculation-of-connectivity-properties">4.2. Calculation of connectivity properties</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculation-of-molecular-structure-descriptors">5. Calculation of Molecular Structure Descriptors</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extended-connectivity-fingerprints-ecfps">5.1. Extended connectivity fingerprints (ECFPs)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fame-atom-environment-fingerprints">5.2. FAME atom environment fingerprints</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#force-field-calculations">6. Force Field Calculations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calculation-of-mmff94-atom-charges">6.1. Calculation of MMFF94 atom charges</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#substructure-searching-and-matching">7. Substructure Searching and Matching</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filtering-molecules-based-on-a-smarts-pattern">7.1. Filtering molecules based on a SMARTS pattern</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cdpl_python_cookbook/index.html">CDPL Python Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cdpl_api_doc/index.html">CDPL API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #aaaaaa" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CDPKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">CDPL Python Tutorial</a></li>
      <li class="breadcrumb-item active"><span class="section-number">1. </span>Molecular Input/Output</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/cdpl_python_tutorial/cdpl_python_tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="molecular-input-output">
<h1><span class="section-number">1. </span>Molecular Input/Output<a class="headerlink" href="#molecular-input-output" title="Link to this heading"></a></h1>
<section id="reading-molecules-from-an-input-file">
<h2><span class="section-number">1.1. </span>Reading molecules from an input file<a class="headerlink" href="#reading-molecules-from-an-input-file" title="Link to this heading"></a></h2>
<p>A frequent task in cheminformatics is to read molecular data from files in different formats.
For this purpose, the CDPL provides data readers for various file formats including <em>SDF</em>, <em>SMILES</em>, <em>MOL</em>, <em>MOL2</em>, <em>CDF</em>, and others.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The CDF format is a binary format and CDPKit’s native file format that can be used to store molecule, reaction, and pharmacophore data in a space and processing time efficient manner.</p>
</div>
<p>The class <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1MoleculeReader.html">Chem.MoleculeReader</a> is rather versatile and can read molecules stored in
any of the supported formats (SDF, SMILES, MOL, MOL2, CDF, etc.) thus eliminating the need to instantiate a format-specific reader implementation.
Here is a simple example of how to read molecules from an input file:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The module <a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1Chem.html">CDPL.Chem</a> provides functionality for the processing of chemical data.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

 <span class="k">def</span> <span class="nf">processMoleculesFromFile</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Reads molecules from the provided input file and outputs the number of atoms and bonds for each molecule.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - input_file (str): Path to the input file.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># Create reader for input molecules (format specified by file extension)</span>
     <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>

     <span class="c1"># create an instance of the default implementation of the Chem.Molecule interface</span>
     <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>

     <span class="c1"># Iterate over each molecule and print atom and bond count</span>
     <span class="k">try</span><span class="p">:</span>
         <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
             <span class="k">try</span><span class="p">:</span>
                 <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Molecule with&#39;</span><span class="p">,</span> <span class="n">mol</span><span class="o">.</span><span class="n">numAtoms</span><span class="p">,</span> <span class="s1">&#39;atoms and&#39;</span><span class="p">,</span> <span class="n">mol</span><span class="o">.</span><span class="n">numBonds</span><span class="p">,</span> <span class="s1">&#39;bonds&#39;</span><span class="p">)</span>
             <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>


 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="n">processMoleculesFromFile</span><span class="p">(</span><span class="s2">&quot;path_to_mol_file.mol2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="retrieving-structure-data-from-mdl-sd-files">
<h2><span class="section-number">1.2. </span>Retrieving structure data from MDL SD-files<a class="headerlink" href="#retrieving-structure-data-from-mdl-sd-files" title="Link to this heading"></a></h2>
<p>Analyzing the data associated with a molecule in an SD-file can be done as shown
in the following example:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can either use <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1MoleculeReader.html">Chem.MoleculeReader</a>
or use the format-specific reader <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1FileSDFMoleculeReader.html">Chem.FileSDFMoleculeReader</a>
molecules from an SD-file. Usually, the more convenient and flexible way to read molecule data is to use Chem.MoleculeReader.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">processStructureDataFromSD</span><span class="p">(</span><span class="n">input_sd_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd"> Retrieves the structure data of each molecule in the provided SD file and outputs it to the console.</span>

<span class="sd"> Parameters:</span>
<span class="sd"> - input_sd_file (str): Path to the input SD file.</span>
<span class="sd"> &quot;&quot;&quot;</span>
     <span class="c1"># Create reader for MDL SD-files</span>
     <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">FileSDFMoleculeReader</span><span class="p">(</span><span class="n">input_sd_file</span><span class="p">)</span>

     <span class="c1"># create an instance of the default implementation of the Chem.Molecule interface</span>
     <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>

     <span class="c1"># Iterate over each molecule in the file and retrieve structure data</span>
     <span class="k">try</span><span class="p">:</span>
         <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
             <span class="k">try</span><span class="p">:</span>
                 <span class="k">if</span> <span class="ow">not</span> <span class="n">Chem</span><span class="o">.</span><span class="n">hasStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error: no structure data available for molecule&#39;</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">))</span>
                     <span class="k">continue</span>

                 <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span> <span class="c1"># retrieve structure data</span>

                 <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Structure data (</span><span class="si">%s</span><span class="s1"> entries) of molecule </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1">:</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_data</span><span class="p">)),</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)))</span>

                 <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">struct_data</span><span class="p">:</span> <span class="c1"># iterate of structure data entries consisting of a header line and the actual data</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Header:&#39;</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">header</span><span class="p">)</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data:&#39;</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
             <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>


 <span class="c1">##########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="n">processStructureDataFromSD</span><span class="p">(</span><span class="s2">&quot;path_to_sd_file.sdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="writing-molecules-to-an-output-file">
<h2><span class="section-number">1.3. </span>Writing molecules to an output file<a class="headerlink" href="#writing-molecules-to-an-output-file" title="Link to this heading"></a></h2>
<p>Once you’ve processed or analyzed your molecules, you may want to save them to an output file.
In the constructor of the class <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1MolecularGraphWriter.html">Chem.MolecularGraphWriter</a> the format of the output
file is determined by its file extension, making it easy to save molecules in different formats such as SDF, MOL, MOL2, CDF, and others without having to write
format-specific code.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Ensure that the file extension you provide matches the standard desired output format. For instance, use <cite>.sdf</cite> for MDL Structure-Data Files, <cite>.mol</cite> for MDL Molfiles, and so on.</p>
</div>
<p>Here’s a simple example of how a list of molecules can be written to an output file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">molsToFiles</span><span class="p">(</span><span class="n">mols</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">],</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Writes a list of molecules to the specified output file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mols (list): List of CDPKit molecules to write to the output file.</span>
<span class="sd">    - output_file (str): Path to the output file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create a writer for the output molecules (format specified by file extension)</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>


 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>
 <span class="n">molsToFiles</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="description-of-structural-atom-environments-as-smiles-strings">
<h2><span class="section-number">1.4. </span>Description of structural atom environments as SMILES strings<a class="headerlink" href="#description-of-structural-atom-environments-as-smiles-strings" title="Link to this heading"></a></h2>
<p>For the extraction of the local chemical environment of an atom in a molecular graph the CDPL provides the utility function
<a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1Chem.html#a342acefab39f6928df4c67c2d86c209a">Chem.getEnvironment</a>.
Fur further processing extracted environments can, e.g., be output as SMILES strings as shown in the following example.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">printAtomEnv</span><span class="p">(</span><span class="n">mols</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts the atom environments of each atom in the provided list of molecules and outputs them as SMILES strings.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mols (list): List of CDPKit molecules to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

        <span class="n">frag</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span>                       <span class="c1"># for storing extracted atom environments</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom environments (radius = 3 bonds)&#39;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">getEnvironment</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">mol</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">frag</span><span class="p">)</span>     <span class="c1"># extract environment of atom reaching out up to three bonds</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveComponents</span><span class="p">(</span><span class="n">frag</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># perceive molecular graph components (required for SMILES generation)</span>

            <span class="n">smiles</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">generateSMILES</span><span class="p">(</span><span class="n">frag</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># generate non-canonical SMILES string with explicit hydrogen atoms</span>

            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Atom #</span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">mol</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)),</span> <span class="n">smiles</span><span class="p">))</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span> <span class="c1"># Example list of BasicMolecules</span>
<span class="n">printAtomEnv</span><span class="p">(</span><span class="n">mols</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="chembl-molecule-standardization-and-parent-structure-extraction">
<h2><span class="section-number">1.5. </span>ChEMBL molecule standardization and parent structure extraction<a class="headerlink" href="#chembl-molecule-standardization-and-parent-structure-extraction" title="Link to this heading"></a></h2>
<p>The CDPL provides a convenient way to standardize molecules using its implementaion of the ChEMBL standardization pipeline <span id="id2">[<a class="reference internal" href="../bibliography.html#id6" title="A. Patrícia Bento, Anne Hersey, Eloy Félix, Greg Landrum, Anna Gaulton, Francis Atkinson, Louisa J. Bellis, Marleen De Veij, and Andrew R. Leach. An open source chemical structure curation pipeline using rdkit. Journal of Cheminformatics, 12(1):51, Sep 2020. URL: https://doi.org/10.1186/s13321-020-00456-1, doi:10.1186/s13321-020-00456-1.">15</a>]</span>.
This process ensures that molecules will be represented in a consistent and standardized manner when they enter downstream processing steps.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">standardize</span><span class="p">(</span><span class="n">chembl_proc</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="p">,</span> <span class="n">in_mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs ChEMBL molecule standardization and parent structure extraction (optional) for a given input molecule using a provided Chem.ChEMBLStandardizer instance.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - chembl_proc (Chem.ChEMBLStandardizer): Instance of the Chem.ChEMBLStandardizer class.</span>
<span class="sd">    - in_mol (Chem.Molecule): Input molecule to standardize.</span>
<span class="sd">    - out_mol (Chem.Molecule): Output molecule to store the standardized molecule.</span>
<span class="sd">    - proc_excluded (bool): If True, molecules flagged as excluded will be processed.</span>
<span class="sd">    - extract_parent (bool): If True, the parent structure will be extracted.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Chem.ChEMBLStandardizer.ChangeFlags: Flags indicating the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># here, the standardization is carried out on a copy of the read input molecule</span>
    <span class="c1"># (if only one molecule instance gets provided as argument, modifications will be made in-place)</span>
    <span class="n">change_flags</span> <span class="o">=</span> <span class="n">chembl_proc</span><span class="o">.</span><span class="n">standardize</span><span class="p">(</span><span class="n">in_mol</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="c1"># perform parent structure extraction (optional)</span>
        <span class="n">change_flags</span> <span class="o">&amp;=</span> <span class="o">~</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span>  <span class="c1"># clear excluded flag possibly set by the standardization</span>
                                                       <span class="c1"># procedure (might change after salt stripping)</span>
        <span class="n">change_flags</span> <span class="o">|=</span> <span class="n">chembl_proc</span><span class="o">.</span><span class="n">getParent</span><span class="p">(</span><span class="n">out_mol</span><span class="p">)</span>     <span class="c1"># extract parent structure (in-place) and add information</span>
                                                       <span class="c1"># about the carried out modifcations</span>
    <span class="k">return</span> <span class="n">change_flags</span>

<span class="k">def</span> <span class="nf">getListOfChangesString</span><span class="p">(</span><span class="n">change_flags</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a string listing the carried out modifications.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - change_flags (Chem.ChEMBLStandardizer.ChangeFlags): Flags indicating the carried out modifications.</span>
<span class="sd">    - verbose (bool): If True, the string will contain a detailed list of the carried out modifications.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: String listing the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">changes</span> <span class="o">=</span> <span class="s1">&#39;   Carried out modifications:&#39;</span>

    <span class="c1"># List of possible changes</span>
    <span class="n">change_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXPLICIT_HYDROGENS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Explicit hydrogens removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">UNKNOWN_STEREO_STANDARDIZED</span><span class="p">,</span> <span class="s1">&#39;Undefined stereocenter information standardized&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">BONDS_KEKULIZED</span><span class="p">,</span> <span class="s1">&#39;Kekule structure generated&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">STRUCTURE_NORMALIZED</span><span class="p">,</span> <span class="s1">&#39;Functional groups normalized&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">CHARGES_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Number of charged atoms reduced&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">TARTRATE_STEREO_CLEARED</span><span class="p">,</span> <span class="s1">&#39;Configuration of chiral tartrate atoms set to undefined&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">STRUCTURE_2D_CORRECTED</span><span class="p">,</span> <span class="s1">&#39;2D structure corrected&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ISOTOPE_INFO_CLEARED</span><span class="p">,</span> <span class="s1">&#39;Isotope information cleared&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">SALT_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Salt components removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">SOLVENT_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Solvent components removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">DUPLICATE_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Duplicate components removed&#39;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">flag</span><span class="p">,</span> <span class="n">description</span> <span class="ow">in</span> <span class="n">change_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">flag</span><span class="p">:</span>
            <span class="n">changes</span> <span class="o">+=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">    * &#39;</span> <span class="o">+</span> <span class="n">description</span>

    <span class="k">return</span> <span class="n">changes</span>

<span class="k">def</span> <span class="nf">getLogMessage</span><span class="p">(</span><span class="n">change_flags</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">mol_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a log message describing the carried out modifications.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - change_flags (Chem.ChEMBLStandardizer.ChangeFlags): Flags indicating the carried out modifications.</span>
<span class="sd">    - proc_excluded (bool): If True, molecules flagged as excluded will be processed.</span>
<span class="sd">    - extract_parent (bool): If True, the parent structure will be extracted.</span>
<span class="sd">    - mol_id (str): Identifier of the molecule.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: Log message describing the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span><span class="p">)</span> <span class="ow">and</span> <span class="n">proc_excluded</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: discarded (flagged as excluded)&#39;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">proc_excluded</span> <span class="ow">and</span> <span class="p">(</span><span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: forwarded unchanged (flagged as excluded)&#39;</span>

    <span class="k">if</span> <span class="n">change_flags</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: modified</span><span class="se">\n</span><span class="si">{</span><span class="n">getListOfChangesString</span><span class="p">(</span><span class="n">change_flags</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: forwarded unchanged&#39;</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="n">chembl_proc</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="p">()</span>

<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="n">in_mol</span> <span class="o">=</span> <span class="n">mol</span>
    <span class="n">out_mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>
    <span class="n">change_flags</span> <span class="o">=</span> <span class="n">standardize</span><span class="p">(</span><span class="n">chembl_proc</span><span class="p">,</span> <span class="n">in_mol</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">extract_parent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">in_mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">or</span> <span class="sa">f</span><span class="s1">&#39;Molecule_</span><span class="si">{</span><span class="n">mols</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
    <span class="n">log_msg</span> <span class="o">=</span> <span class="n">getLogMessage</span><span class="p">(</span><span class="n">change_flags</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">extract_parent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mol_id</span><span class="o">=</span><span class="n">mol_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">log_msg</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="standardization-prediction-of-protonation-states">
<h2><span class="section-number">1.6. </span>Standardization/prediction of protonation states<a class="headerlink" href="#standardization-prediction-of-protonation-states" title="Link to this heading"></a></h2>
<p>In the realm of computational chemistry and molecular modeling, understanding and predicting the behavior of molecules
often hinges on the finer details. One such critical detail is the protonation state of the functional groups in a molecule.
Protonation states dictate formal charges of atoms and have an impact on molecular 3D structure and reactivity. Knowing the
pre-dominant protonation states of functional groups under certain condition is therefore vital for methods predicting
drug binding, enzymatic reactions, etc.</p>
<p>The class <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1ProtonationStateStandardizer.html">Chem.ProtonationStateStandardizer</a>
implements several protonation state generation algorithms.
The following code snippet shows how to generate protonation/formal charge states of acidic/basic functional groups that are likely
to occur under physiological conditions.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create and initialize an instance of the class Chem.ProtonationStateStandardizer which</span>
<span class="c1"># implements the protonation state generation algorithm</span>
<span class="n">prot_state_gen</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ProtonationStateStandardizer</span><span class="p">()</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="c1"># compose a simple molecule identifier</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># protonate/deprotonate functional groups for phys. conditions</span>
        <span class="n">prot_state_gen</span><span class="o">.</span><span class="n">standardize</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ProtonationStateStandardizer</span><span class="o">.</span><span class="n">PHYSIOLOGICAL_CONDITION_STATE</span><span class="p">)</span>

        <span class="c1"># enforce an update of the molecule components list (structure might have changed)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveComponents</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing or output of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="d-structure-and-conformer-ensemble-generation">
<h1><span class="section-number">2. </span>3D Structure and Conformer Ensemble Generation<a class="headerlink" href="#d-structure-and-conformer-ensemble-generation" title="Link to this heading"></a></h1>
<p>In chemistry, conformational isomerism is a form of stereoisomerism in which the isomers can be interconverted just by rotations about formally single
bonds. While any two arrangements of atoms in a molecule that differ by rotation about single bonds can be referred to as different conformations,
conformations that correspond to local minima on the potential energy surface are specifically called conformational isomers or conformers.</p>
<p>In the realm of cheminformatics and molecular modeling many methods and algorithms require 3D structures or even whole
conformer ensembles as input.
For a seamless integration with such methods CDPKit provides functionality in package <a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1ConfGen.html">CDPL.ConfGen</a> that allows
to generate both single low-energy 3D structures and diverse conformer ensembles solely from connection table information.
The high-quality conformer ensemble generation tool included in CDPKit is called <em>CONFORGE</em>. For details on its implementation and performance
see <span id="id3">[<a class="reference internal" href="../bibliography.html#id12" title="Thomas Seidel, Christian Permann, Oliver Wieder, Stefan M. Kohlbacher, and Thierry Langer. High-quality conformer generation with conforge: algorithm and performance assessment. Journal of Chemical Information and Modeling, 63(17):5549-5570, 2023. PMID: 37624145. URL: https://doi.org/10.1021/acs.jcim.3c00563, arXiv:https://doi.org/10.1021/acs.jcim.3c00563, doi:10.1021/acs.jcim.3c00563.">5</a>]</span>.</p>
<section id="generating-single-low-energy-3d-structures">
<h2><span class="section-number">2.1. </span>Generating single low-energy 3D structures<a class="headerlink" href="#generating-single-low-energy-3d-structures" title="Link to this heading"></a></h2>
<p>Here’s how you can generate a single low-energy 3D structure for a molecule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ConfGen</span> <span class="k">as</span> <span class="nn">ConfGen</span>

 <span class="k">def</span> <span class="nf">generate3dConformation</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">struct_gen</span><span class="p">:</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">StructureGenerator</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Generates a low-energy 3D structure of the argument molecule using the provided initialized ConfGen.StructureGenerator instance.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - mol (Chem.Molecule): Molecule to generate a 3D structure for.</span>
<span class="sd">     - struct_gen (ConfGen.StructureGenerator): Instance of the ConfGen.StructureGenerator class.</span>

<span class="sd">     Returns:</span>
<span class="sd">     - int: Status code indicating the success of the 3D structure generation.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># prepare the molecule for 3D structure generation</span>
     <span class="n">ConfGen</span><span class="o">.</span><span class="n">prepareForConformerGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># generate the 3D structure</span>
     <span class="n">status</span> <span class="o">=</span> <span class="n">struct_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># if sucessful, store the generated conformer ensemble as</span>
     <span class="c1"># per atom 3D coordinates arrays (= the way conformers are represented in CDPKit)</span>
     <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span><span class="p">:</span>
         <span class="n">struct_gen</span><span class="o">.</span><span class="n">setCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># return status code</span>
     <span class="k">return</span> <span class="n">status</span>


 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="n">max_time</span> <span class="o">=</span> <span class="mi">3600</span> <span class="c1"># Max. allowed molecule processing time in seconds (default: 3600 sec)</span>

 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

 <span class="c1"># create writer for the generated 3D structures (format specified by file extension)</span>
 <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

 <span class="c1"># export only a single 3D structure (in case of multi-conf. input molecules)</span>
 <span class="n">Chem</span><span class="o">.</span><span class="n">setMultiConfExportParameter</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

 <span class="c1"># create and initialize an instance of the class ConfGen.StructureGenerator which will</span>
 <span class="c1"># perform the actual 3D structure generation work</span>
 <span class="n">struct_gen</span> <span class="o">=</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">StructureGenerator</span><span class="p">()</span>

 <span class="n">struct_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">timeout</span> <span class="o">=</span> <span class="n">max_time</span> <span class="o">*</span> <span class="mi">1000</span> <span class="c1"># apply the -t argument</span>

 <span class="c1"># dictionary mapping status codes to human readable strings</span>
 <span class="n">status_to_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">UNINITIALIZED</span>                  <span class="p">:</span> <span class="s1">&#39;uninitialized&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TIMEOUT</span>                        <span class="p">:</span> <span class="s1">&#39;max. processing time exceeded&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">ABORTED</span>                        <span class="p">:</span> <span class="s1">&#39;aborted&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_SETUP_FAILED</span>        <span class="p">:</span> <span class="s1">&#39;force field setup failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_MINIMIZATION_FAILED</span> <span class="p">:</span> <span class="s1">&#39;force field structure refinement failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_LIBRARY_NOT_SET</span>       <span class="p">:</span> <span class="s1">&#39;fragment library not available&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_FAILED</span>       <span class="p">:</span> <span class="s1">&#39;fragment conformer generation failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_TIMEOUT</span>      <span class="p">:</span> <span class="s1">&#39;fragment conformer generation timeout&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_ALREADY_PROCESSED</span>     <span class="p">:</span> <span class="s1">&#39;fragment already processed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TORSION_DRIVING_FAILED</span>         <span class="p">:</span> <span class="s1">&#39;torsion driving failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">CONF_GEN_FAILED</span>                <span class="p">:</span> <span class="s1">&#39;conformer generation failed&#39;</span> <span class="p">}</span>

 <span class="c1"># process molecules one after the other</span>
 <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
     <span class="c1"># compose a simple molecule identifier</span>
     <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

     <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
     <span class="k">else</span><span class="p">:</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

     <span class="k">try</span><span class="p">:</span>
         <span class="c1"># generate 3D structure of the read molecule</span>
         <span class="n">status</span> <span class="o">=</span> <span class="n">generate3dConformation</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">struct_gen</span><span class="p">)</span>

         <span class="c1"># check for severe error reported by status code</span>
         <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span><span class="p">:</span>
             <span class="c1"># enforce the output of 3D coordinates in case of MDL file formats</span>
             <span class="n">Chem</span><span class="o">.</span><span class="n">setMDLDimensionality</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>

             <span class="c1"># output the generated 3D structure</span>
             <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing 3D structure of molecule </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: 3D structure generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

 <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="generating-conformer-ensembles">
<h2><span class="section-number">2.2. </span>Generating conformer ensembles<a class="headerlink" href="#generating-conformer-ensembles" title="Link to this heading"></a></h2>
<p>For some applications it is necessary to generate multiple conformations for a molecule, e.g. to assess its flexibility or investigate binding
capabilities towards a particular target receptor.</p>
<p>The function <code class="docutils literal notranslate"><span class="pre">generate_conformation_ensembles()</span></code> generates a conformer ensemble for a given molecule using an
instance of class <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1ConfGen_1_1ConformerGenerator.html">ConfGen.ConformerGenerator</a>.
The parameters <em>min_rmsd</em>, <em>e_window</em>, and <em>max_confs</em> are used to control the generated conformer ensemble’s diversity, energy and size.
They can be initialized at the beginning of the example script with the desired values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ConfGen</span> <span class="k">as</span> <span class="nn">ConfGen</span>

 <span class="k">def</span> <span class="nf">generateConformationEnsembles</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">,</span> <span class="n">conf_gen</span><span class="p">:</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ConformerGenerator</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
<span class="w">     </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Generates a conformation ensemble for the argument molecule using the provided initialized ConfGen.ConformerGenerator instance.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - mol (Chem.BasicMolecule): Molecule to generate a conformation ensemble for.</span>
<span class="sd">     - conf_gen (ConfGen.ConformerGenerator): Instance of the ConfGen.ConformerGenerator class.</span>

<span class="sd">     Returns:</span>
<span class="sd">     - int: Status code indicating the success of the conformation ensemble generation.</span>
<span class="sd">     - int: Number of generated conformers.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># prepare the molecule for conformer generation</span>
     <span class="n">ConfGen</span><span class="o">.</span><span class="n">prepareForConformerGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># generate the conformer ensemble</span>
     <span class="n">status</span> <span class="o">=</span> <span class="n">conf_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
     <span class="n">num_confs</span> <span class="o">=</span> <span class="n">conf_gen</span><span class="o">.</span><span class="n">getNumConformers</span><span class="p">()</span>

     <span class="c1"># if sucessful, store the generated conformer ensemble as</span>
     <span class="c1"># per atom 3D coordinates arrays (= the way conformers are represented in CDPKit)</span>
     <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span> <span class="ow">or</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TOO_MUCH_SYMMETRY</span><span class="p">:</span>
         <span class="n">conf_gen</span><span class="o">.</span><span class="n">setConformers</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
     <span class="k">else</span><span class="p">:</span>
         <span class="n">num_confs</span> <span class="o">=</span> <span class="mi">0</span>

     <span class="c1"># return status code and the number of generated conformers</span>
     <span class="k">return</span> <span class="p">(</span><span class="n">status</span><span class="p">,</span> <span class="n">num_confs</span><span class="p">)</span>


 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="c1"># Settings</span>
 <span class="n">max_time</span> <span class="o">=</span> <span class="mi">3600</span> <span class="c1"># Max. allowed molecule processing time in seconds (default: 3600 sec)</span>
 <span class="n">min_rmsd</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="c1"># Output conformer RMSD threshold (default: 0.5)</span>
 <span class="n">e_window</span> <span class="o">=</span> <span class="mf">20.0</span> <span class="c1"># Output conformer energy window (default: 20.0)</span>
 <span class="n">max_confs</span> <span class="o">=</span> <span class="mi">100</span> <span class="c1"># Max. output ensemble size (default: 100)</span>

 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

 <span class="c1"># create writer for the generated conformer ensembles (format specified by file extension)</span>
 <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

 <span class="c1"># create and initialize an instance of the class ConfGen.ConformerGenerator which</span>
 <span class="c1"># will perform the actual conformer ensemble generation work</span>
 <span class="n">conf_gen</span> <span class="o">=</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ConformerGenerator</span><span class="p">()</span>

 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">timeout</span> <span class="o">=</span> <span class="n">max_time</span> <span class="o">*</span> <span class="mi">1000</span>          <span class="c1"># apply the -t argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">minRMSD</span> <span class="o">=</span> <span class="n">min_rmsd</span>                 <span class="c1"># apply the -r argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">energyWindow</span> <span class="o">=</span> <span class="n">e_window</span>            <span class="c1"># apply the -e argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">maxNumOutputConformers</span> <span class="o">=</span> <span class="n">max_confs</span> <span class="c1"># apply the -n argument</span>

 <span class="c1"># dictionary mapping status codes to human readable strings</span>
 <span class="n">status_to_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">UNINITIALIZED</span>                  <span class="p">:</span> <span class="s1">&#39;uninitialized&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TIMEOUT</span>                        <span class="p">:</span> <span class="s1">&#39;max. processing time exceeded&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">ABORTED</span>                        <span class="p">:</span> <span class="s1">&#39;aborted&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_SETUP_FAILED</span>        <span class="p">:</span> <span class="s1">&#39;force field setup failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_MINIMIZATION_FAILED</span> <span class="p">:</span> <span class="s1">&#39;force field structure refinement failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_LIBRARY_NOT_SET</span>       <span class="p">:</span> <span class="s1">&#39;fragment library not available&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_FAILED</span>       <span class="p">:</span> <span class="s1">&#39;fragment conformer generation failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_TIMEOUT</span>      <span class="p">:</span> <span class="s1">&#39;fragment conformer generation timeout&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_ALREADY_PROCESSED</span>     <span class="p">:</span> <span class="s1">&#39;fragment already processed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TORSION_DRIVING_FAILED</span>         <span class="p">:</span> <span class="s1">&#39;torsion driving failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">CONF_GEN_FAILED</span>                <span class="p">:</span> <span class="s1">&#39;conformer generation failed&#39;</span> <span class="p">}</span>


 <span class="c1"># process molecules one after the other</span>
 <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
     <span class="c1"># compose a simple molecule identifier</span>
     <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

     <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
     <span class="k">else</span><span class="p">:</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

     <span class="k">try</span><span class="p">:</span>
         <span class="c1"># generate conformer ensemble for read molecule</span>
         <span class="n">status</span><span class="p">,</span> <span class="n">num_confs</span> <span class="o">=</span> <span class="n">generateConformationEnsembles</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">conf_gen</span><span class="p">)</span>

         <span class="c1"># output generated ensemble (if available)</span>
         <span class="k">if</span> <span class="n">num_confs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
             <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: output of conformer ensemble for molecule </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: conformer ensemble generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

 <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="pharmacophore-generation-and-processing">
<h1><span class="section-number">3. </span>Pharmacophore Generation and Processing<a class="headerlink" href="#pharmacophore-generation-and-processing" title="Link to this heading"></a></h1>
<p>This section is about the generation and processing of pharmacophore models.
The pharmacophore concept is widely used in drug design and cheminformatics and represents a versatile tool to understand and
describe interactions between ligands and their biological targets.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Functionality for the generation and processing of pharmacophore models resides in package
<a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1Pharm.html">CDPL.Pharm</a>.
Currently available pharmacophore input/output formats are <a class="reference external" href="https://www.inteligand.com/ligandscout">LigandScout’s</a> <em>PML</em> and
CDPKit’s native <em>CDF</em> format.</p>
</div>
<section id="pharmacophore-generation-for-single-conformer-molecules">
<h2><span class="section-number">3.1. </span>Pharmacophore generation for single-conformer molecules<a class="headerlink" href="#pharmacophore-generation-for-single-conformer-molecules" title="Link to this heading"></a></h2>
<p>The following example script generates a pharmacophore model for each input molecule
and outputs the pharmacophore data in PML format.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># first call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span> <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4_name</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                    <span class="c1"># use the name of the input molecule as pharmacophore name</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>          <span class="c1"># generate the pharmacophore</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">setName</span><span class="p">(</span><span class="n">ph4</span><span class="p">,</span> <span class="n">ph4_name</span><span class="p">)</span>        <span class="c1"># set the pharmacophore name</span>

    <span class="k">return</span> <span class="n">ph4</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create writer for the generated pharmacophores (format specified by file extension)</span>
<span class="n">writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.pml&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="c1"># compose a simple molecule identifier</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>         <span class="c1"># generate pharmacophore</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>   <span class="c1"># output pharmacophore</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing generated pharmacophore </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
       <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="pharmacophore-generation-for-multi-conformer-molecules">
<h2><span class="section-number">3.2. </span>Pharmacophore generation for multi-conformer molecules<a class="headerlink" href="#pharmacophore-generation-for-multi-conformer-molecules" title="Link to this heading"></a></h2>
<p>In the following example script a pharmacophore model for each conformer of the input molecule will be generated.
The name of the generated pharmacophore is set to the name of the corresponding molecule plus a suffix specifying the
conformer index.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">conf_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the molecule using atom coordinates of the specified conformation.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>
<span class="sd">    - conf_idx (int): Index of the conformation to use for the pharmacophore generation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule using coordinates of the specifies conformer.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">conf_idx</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>                                    <span class="c1"># for a new molecule</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># first call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span>     <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                    <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4_name</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                        <span class="c1"># use the name of the input molecule as pharmacophore name</span>

    <span class="c1"># use atom 3D coordinates of the specified conf.</span>
    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">setAtom3DCoordinatesFunction</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">AtomConformer3DCoordinatesFunctor</span><span class="p">(</span><span class="n">conf_idx</span><span class="p">))</span>

    <span class="c1"># append conformer index to the pharmacophore name</span>
    <span class="n">ph4_name</span> <span class="o">+=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">conf_idx</span><span class="p">)</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>          <span class="c1"># generate the pharmacophore</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">setName</span><span class="p">(</span><span class="n">ph4</span><span class="p">,</span> <span class="n">ph4_name</span><span class="p">)</span>        <span class="c1"># set the pharmacophore name</span>

    <span class="k">return</span> <span class="n">ph4</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create writer for the generated pharmacophores (format specified by file extension)</span>
<span class="n">writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.pml&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
   <span class="c1"># compose a simple molecule identifier</span>
   <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

   <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
       <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
   <span class="k">else</span><span class="p">:</span>
       <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

   <span class="n">num_confs</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getNumConformations</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

   <span class="k">try</span><span class="p">:</span>
       <span class="k">for</span> <span class="n">conf_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_confs</span><span class="p">):</span>               <span class="c1"># for each conformer</span>
           <span class="n">ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">conf_idx</span><span class="p">)</span>  <span class="c1"># generate pharmacophore</span>

           <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>   <span class="c1"># output pharmacophore</span>
               <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing generated pharmacophore </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

   <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
       <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="generating-ligand-receptor-interaction-pharmacophores">
<h2><span class="section-number">3.3. </span>Generating ligand-receptor interaction pharmacophores<a class="headerlink" href="#generating-ligand-receptor-interaction-pharmacophores" title="Link to this heading"></a></h2>
<p>The following example shows how to generate 3D pharmacophore models that describe observed interactions between
a molecule and proximal binding site residues.
The script also demonstrates how to read and preprocess biological macromolecules.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The receptor structure can be provided in the formats *.mol2, *.pdb, or *.mmtf</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">processReceptorStructure</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strip_res_list</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads and preprocesses the specified receptor structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - path (str): Path to the receptor structure file.</span>
<span class="sd">    - strip_res_list (bool): Whitespace separated list of PDB three-letter codes specifying residues to remove from the receptor structure (e.g. an existing ligand).</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Chem.Molecule: Receptor structure.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># create reader for receptor structure (format specified by file extension)</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="s2">&quot;path_to_receptor_structure_file.pdb&quot;</span><span class="p">)</span><span class="s2">&quot;)</span>

    <span class="n">sup_fmts</span> <span class="o">=</span> <span class="p">[</span> <span class="n">Chem</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MOL2</span><span class="p">,</span>
                <span class="n">Biomol</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">PDB</span><span class="p">,</span>
                <span class="n">Biomol</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MMTF</span> <span class="p">]</span>

    <span class="c1"># check if the format is supported by this script</span>
    <span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">getDataFormat</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sup_fmts</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: receptor input file format </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> not supported&#39;</span> <span class="o">%</span> <span class="n">name_and_ext</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="n">rec_mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>    <span class="c1"># create an instance of the default implementation of the</span>
                                      <span class="c1"># Chem.Molecule interface that will store the receptor struct.</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">):</span>  <span class="c1"># read receptor structure</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading receptor structure failed&#39;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading receptor structure failed:</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="c1"># preprocess the receptor structure (removal of residues and</span>
    <span class="c1"># calculation of properties required by the pharm. generation procedure)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if structure comes from an MOL2 file, convert MOL2 residue data into PDB-style data</span>
        <span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">getDataFormat</span><span class="p">()</span> <span class="o">==</span> <span class="n">Chem</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MOL2</span><span class="p">:</span>
            <span class="n">Biomol</span><span class="o">.</span><span class="n">convertMOL2ToPDBResidueInfo</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

        <span class="n">rem_atoms</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># delete atoms belonging to residues that should be stripped</span>
        <span class="k">if</span> <span class="n">strip_res_list</span><span class="p">:</span>
            <span class="n">atoms_to_rem</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span> <span class="c1"># will store the atoms to delete</span>
            <span class="n">res_to_strip</span> <span class="o">=</span> <span class="p">{</span> <span class="n">tlc</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">tlc</span> <span class="ow">in</span> <span class="n">strip_res_list</span> <span class="p">}</span>

            <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">rec_mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>     <span class="c1"># identify and note atoms belonging to the stripped residues</span>
                <span class="k">if</span> <span class="n">Biomol</span><span class="o">.</span><span class="n">getResidueCode</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">res_to_strip</span><span class="p">:</span>
                    <span class="n">atoms_to_rem</span><span class="o">.</span><span class="n">addAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">atoms_to_rem</span><span class="o">.</span><span class="n">numAtoms</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">rec_mol</span> <span class="o">-=</span> <span class="n">atoms_to_rem</span>    <span class="c1"># delete atoms from the receptor structure</span>
                <span class="n">rem_atoms</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># prepares the receptor structure for pharmacophore generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">):</span>                    <span class="c1"># make implicit hydrogens (if any) explicit</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">calcHydrogen3DCoordinates</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">)</span>               <span class="c1"># calculate 3D coordinates for the added expl. hydrogens</span>
            <span class="n">Biomol</span><span class="o">.</span><span class="n">setHydrogenResidueSequenceInfo</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># set residue information for the added expl. hydrogens</span>

        <span class="n">MolProp</span><span class="o">.</span><span class="n">calcAtomHydrophobicities</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>          <span class="c1"># calculate atom hydrophobicity values (needed for hydrophobic</span>
                                                                <span class="c1"># pharm. feature generation)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of receptor structure failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">rec_mol</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># Settings</span>
<span class="n">strip_res_list</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># Whitespace separated list of PDB three-letter codes specifying residues to remove from the receptor structure (e.g. an existing ligand)</span>
<span class="n">gen_x_vols</span> <span class="o">=</span> <span class="kc">False</span>     <span class="c1"># Generate exclusion volume spheres on pharm. feature atoms of interacting residues</span>


<span class="n">lig_mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of ligand molecules</span>

<span class="n">rec_mol</span> <span class="o">=</span> <span class="n">processReceptorStructure</span><span class="p">(</span><span class="n">path_to_receptor_structure_file</span><span class="p">,</span> <span class="n">strip_res_list</span><span class="p">)</span>  <span class="c1"># read and preprocess the receptor structure</span>
<span class="n">ph4_writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_pha_file.pml&quot;</span><span class="p">)</span>                    <span class="c1"># create writer for the generated pharmacophores</span>
                                                                                     <span class="c1"># (format specified by file extension)</span>

<span class="n">ia_ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>     <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore</span>
                                        <span class="c1"># interface that will store the generated pharmacophores</span>

<span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">InteractionPharmacophoreGenerator</span><span class="p">()</span> <span class="c1"># create an instance of the pharmacophore generator</span>

<span class="n">ph4_gen</span><span class="o">.</span><span class="n">addExclusionVolumes</span><span class="p">(</span><span class="n">gen_x_vols</span><span class="p">)</span> <span class="c1"># specify whether to generate exclusion volume spheres</span>
                                        <span class="c1"># on pharm. feature atoms of interacting residues</span>

<span class="c1"># process ligand molecules one after the other</span>
<span class="k">for</span> <span class="n">lig_mol</span> <span class="ow">in</span> <span class="n">lig_mols</span><span class="p">:</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="c1"># compose a simple ligand identifier for messages</span>

    <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>  <span class="c1"># fallback if name is empty or not available</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">)</span> <span class="c1"># make ligand ready for pharm. generation</span>

        <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">,</span> <span class="n">rec_mol</span><span class="p">,</span> <span class="n">ia_ph4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> <span class="c1"># generate the pharmacophore (True = extract ligand environment residues on-the-fly)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">ph4_writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ia_ph4</span><span class="p">):</span> <span class="c1"># output pharmacophore</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing interaction pharmacophore of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>               <span class="c1"># handle exception raised in case of severe write errors</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing interaction pharmacophore of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>                   <span class="c1"># handle exception raised in case of severe processing errors</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: interaction pharmacophore generation for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">ph4_writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="alignment-of-molecules-to-a-reference-pharmacophore">
<h2><span class="section-number">3.4. </span>Alignment of molecules to a reference pharmacophore<a class="headerlink" href="#alignment-of-molecules-to-a-reference-pharmacophore" title="Link to this heading"></a></h2>
<p>The following example shows how to align a set of molecules to a reference pharmacophore.</p>
<p>The function <code class="docutils literal notranslate"><span class="pre">readRefPharmacophore()</span></code> reads the reference pharmacophore from a specified file.
The function <code class="docutils literal notranslate"><span class="pre">genPharmacophore()</span></code> generates and returns the pharmacophore of a molecule.
The function <code class="docutils literal notranslate"><span class="pre">clearFeatureOrientations()</span></code> removes feature orientation information and sets the feature geometry to <code class="docutils literal notranslate"><span class="pre">Pharm.FeatureGeometry.SPHERE</span></code>.</p>
<p>The following variables control the alignment process and the reported results:</p>
<ul class="simple">
<li><p><em>pos_only</em>: Controls whether only the position of features is considered during alignment</p></li>
<li><p><em>min_pose_rmsd</em>: Controls the minimum required RMSD between two consecutively output molecule alignment poses</p></li>
<li><p><em>num_out_almnts</em>: Is used to control the number of top-ranked alignment solutions to output per molecule (default: best alignment solution only)</p></li>
<li><p><em>exhaustive</em>: Specifies whether an exhaustive alignment search should be performed</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">readRefPharmacophore</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads and returns the specified alignment reference pharmacophore.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - filename (str): Name of the file storing the reference pharmacophore.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Reference pharmacophore.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># create pharmacophore reader instance</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreReader</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

    <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span> <span class="c1"># read reference pharmacophore</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading reference pharmacophore failed&#39;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading reference pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the given molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>       <span class="c1"># call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span>    <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                   <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>                         <span class="c1"># generate the pharmacophore</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="k">def</span> <span class="nf">clearFeatureOrientations</span><span class="p">(</span><span class="n">ph4</span><span class="p">:</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Removes feature orientation informations and sets the feature geometry to Pharm.FeatureGeometry.SPHERE.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ph4 (Pharm.BasicPharmacophore): Pharmacophore to edit.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">ftr</span> <span class="ow">in</span> <span class="n">ph4</span><span class="p">:</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">clearOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">)</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">setGeometry</span><span class="p">(</span><span class="n">ftr</span><span class="p">,</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">SPHERE</span><span class="p">)</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># Settings</span>
<span class="n">pos_only</span> <span class="o">=</span> <span class="kc">True</span>     <span class="c1"># True = only position of features is considered during alignment</span>
<span class="n">num_out_almnts</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># Number of top-ranked alignment solutions to output per molecule (default: best alignment solution only)</span>
<span class="n">min_pose_rmsd</span> <span class="o">=</span> <span class="mf">0.0</span> <span class="c1"># Minimum required RMSD between two consecutively output molecule alignment poses</span>
<span class="n">exhaustive</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># Perform an exhaustive alignment search (default: False)</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># read the reference pharmacophore</span>
<span class="n">ref_ph4</span> <span class="o">=</span> <span class="n">readRefPharmacophore</span><span class="p">(</span><span class="s2">&quot;path_to_reference_pharmacophore.pml&quot;</span><span class="p">)</span>

<span class="c1"># create writer for aligned molecules (format specified by file extension)</span>
<span class="n">mol_writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

<span class="c1"># create instance of class implementing the pharmacophore alignment algorithm</span>
<span class="n">almnt</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreAlignment</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># True = aligned features have to be within the tolerance spheres of the ref. features</span>

<span class="k">if</span> <span class="n">pos_only</span><span class="p">:</span>                          <span class="c1"># clear feature orientation information</span>
    <span class="n">clearFeatureOrientations</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">)</span>

<span class="n">almnt</span><span class="o">.</span><span class="n">addFeatures</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>               <span class="c1"># set reference features (True = first set = reference)</span>
<span class="n">almnt</span><span class="o">.</span><span class="n">performExhaustiveSearch</span><span class="p">(</span><span class="n">exhaustive</span><span class="p">)</span> <span class="c1"># set minimum number of top. mapped feature pairs</span>

<span class="c1"># create pharmacophore fit score calculator instance</span>
<span class="n">almnt_score</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreFitScore</span><span class="p">()</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="c1"># compose a simple molecule identifier</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>  <span class="c1"># fallback if name is empty</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">mol_ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># generate input molecule pharmacophore</span>

        <span class="k">if</span> <span class="n">pos_only</span><span class="p">:</span>                            <span class="c1"># clear feature orientation information</span>
            <span class="n">clearFeatureOrientations</span><span class="p">(</span><span class="n">mol_ph4</span><span class="p">)</span>

        <span class="n">almnt</span><span class="o">.</span><span class="n">clearEntities</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>         <span class="c1"># clear features of previously aligned pharmacophore</span>
        <span class="n">almnt</span><span class="o">.</span><span class="n">addFeatures</span><span class="p">(</span><span class="n">mol_ph4</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># specify features of the pharmacophore to align</span>

        <span class="n">almnt_solutions</span> <span class="o">=</span> <span class="p">[]</span>               <span class="c1"># stores the found alignment solutions</span>

        <span class="k">while</span> <span class="n">almnt</span><span class="o">.</span><span class="n">nextAlignment</span><span class="p">():</span>                                     <span class="c1"># iterate over all alignment solutions that can be found</span>
            <span class="n">score</span> <span class="o">=</span> <span class="n">almnt_score</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">,</span> <span class="n">mol_ph4</span><span class="p">,</span> <span class="n">almnt</span><span class="o">.</span><span class="n">getTransform</span><span class="p">())</span>  <span class="c1"># calculate alignment score</span>
            <span class="n">xform</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Matrix4D</span><span class="p">(</span><span class="n">almnt</span><span class="o">.</span><span class="n">getTransform</span><span class="p">())</span>                  <span class="c1"># make a copy of the alignment transformation (mol. ph4 -&gt; ref. ph4)</span>

            <span class="n">almnt_solutions</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">score</span><span class="p">,</span> <span class="n">xform</span><span class="p">))</span>

        <span class="n">saved_coords</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Vector3DArray</span><span class="p">()</span>      <span class="c1"># create data structure for storing 3D coordinates</span>

        <span class="n">Chem</span><span class="o">.</span><span class="n">get3DCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">saved_coords</span><span class="p">)</span> <span class="c1"># save the original atom coordinates</span>

        <span class="n">struct_data</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">hasStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>           <span class="c1"># get existing structure data if available</span>
            <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>                                    <span class="c1"># otherwise create and set new structure data</span>
            <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">StringDataBlock</span><span class="p">()</span>

            <span class="n">Chem</span><span class="o">.</span><span class="n">setStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">strut</span><span class="p">)</span>

        <span class="c1"># add alignment score entry to struct. data</span>
        <span class="n">struct_data</span><span class="o">.</span><span class="n">addEntry</span><span class="p">(</span><span class="s1">&#39;&lt;PharmFitScore&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

        <span class="n">output_cnt</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">last_pose</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># order solutions by desc. alignment score</span>
        <span class="n">almnt_solutions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">almnt_solutions</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">entry</span><span class="p">:</span> <span class="n">entry</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># output molecule alignment poses until the max. number of best output solutions has been reached</span>
        <span class="k">for</span> <span class="n">solution</span> <span class="ow">in</span> <span class="n">almnt_solutions</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">output_cnt</span> <span class="o">==</span> <span class="n">num_out_almnts</span><span class="p">:</span>
                <span class="k">break</span>

            <span class="n">curr_pose</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Vector3DArray</span><span class="p">(</span><span class="n">saved_coords</span><span class="p">)</span>

            <span class="n">Math</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">curr_pose</span><span class="p">,</span> <span class="n">solution</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># transform atom coordinates</span>

            <span class="c1"># check whether the current pose is &#39;different enough&#39; from</span>
            <span class="c1"># the last pose to qualify for output</span>
            <span class="k">if</span> <span class="n">min_pose_rmsd</span> <span class="o">&gt;</span> <span class="mf">0.0</span> <span class="ow">and</span> <span class="n">last_pose</span> <span class="ow">and</span> <span class="n">Math</span><span class="o">.</span><span class="n">calcRMSD</span><span class="p">(</span><span class="n">last_pose</span><span class="p">,</span> <span class="n">curr_pose</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_pose_rmsd</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># apply the transformed atom coordinates</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">set3DCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">curr_pose</span><span class="p">)</span>

            <span class="c1"># store alignment score in the struct. data entry</span>
            <span class="n">struct_data</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_data</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">setData</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">solution</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;.4f&#39;</span><span class="p">))</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">mol_writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span> <span class="c1"># output the alignment pose of the molecule</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing alignment pose of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe write errors</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing alignment pose of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

            <span class="n">last_pose</span> <span class="o">=</span> <span class="n">curr_pose</span>
            <span class="n">output_cnt</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore alignment of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">mol_writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="retrieving-information-about-pharmacophores-and-features">
<h2><span class="section-number">3.5. </span>Retrieving information about pharmacophores and features<a class="headerlink" href="#retrieving-information-about-pharmacophores-and-features" title="Link to this heading"></a></h2>
<p>The script below demonstrates how to retrieve basic properties of pharmacophore features.
The function <code class="docutils literal notranslate"><span class="pre">print_pharmacophore_properties()</span></code> outputs all (available) properties of the features stored
in the given feature container. Built-in feature properties are feature type, geometry, tolerance, weight, and
hydrophobicity.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">print_pharmacophore_properties</span><span class="p">(</span><span class="n">ph4</span><span class="p">:</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainer</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs all (available) properties of the features stored in the given feature container.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ph4 (Pharm.FeatureContainer): Feature container to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ftr_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">UNKNOWN</span>               <span class="p">:</span> <span class="s1">&#39;UNKNOWN&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HYDROPHOBIC</span>           <span class="p">:</span> <span class="s1">&#39;HYDROPHOBIC&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">AROMATIC</span>              <span class="p">:</span> <span class="s1">&#39;AROMATIC&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">NEGATIVE_IONIZABLE</span>    <span class="p">:</span> <span class="s1">&#39;NEGATIVE_IONIZABLE&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">POSITIVE_IONIZABLE</span>    <span class="p">:</span> <span class="s1">&#39;POSITIVE_IONIZABLE&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">H_BOND_DONOR</span>          <span class="p">:</span> <span class="s1">&#39;H_BOND_DONOR&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">H_BOND_ACCEPTOR</span>       <span class="p">:</span> <span class="s1">&#39;H_BOND_ACCEPTOR&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HALOGEN_BOND_DONOR</span>    <span class="p">:</span> <span class="s1">&#39;HALOGEN_BOND_DONOR&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HALOGEN_BOND_ACCEPTOR</span> <span class="p">:</span> <span class="s1">&#39;HALOGEN_BOND_ACCEPTOR&#39;</span><span class="p">,</span>
                     <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">EXCLUSION_VOLUME</span>      <span class="p">:</span> <span class="s1">&#39;EXCLUSION_VOLUME&#39;</span> <span class="p">}</span>

    <span class="n">geom_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">UNDEF</span>   <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                 <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">SPHERE</span>  <span class="p">:</span> <span class="s1">&#39;SPHERE&#39;</span><span class="p">,</span>
                 <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">VECTOR</span>  <span class="p">:</span> <span class="s1">&#39;VECTOR&#39;</span><span class="p">,</span>
                 <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">PLANE</span>   <span class="p">:</span> <span class="s1">&#39;PLANE&#39;</span> <span class="p">}</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Composition of pharmacophore </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1">:&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">ph4</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ph4</span><span class="p">)):</span>
        <span class="n">ftr</span> <span class="o">=</span> <span class="n">ph4</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39; - Feature #</span><span class="si">%s</span><span class="s1">:&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">ftr_type_str</span><span class="p">[</span><span class="n">Pharm</span><span class="o">.</span><span class="n">getType</span><span class="p">(</span><span class="n">ftr</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Geometry: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geom_str</span><span class="p">[</span><span class="n">Pharm</span><span class="o">.</span><span class="n">getGeometry</span><span class="p">(</span><span class="n">ftr</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Tolerance: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getTolerance</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Weight: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getWeight</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Optional: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getOptionalFlag</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Disabled: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getDisabledFlag</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Length: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getLength</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Hydrophobicity: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getHydrophobicity</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">has3DCoordinates</span><span class="p">(</span><span class="n">ftr</span><span class="p">):</span>         <span class="c1"># Pharm.Feature derives from Chem.Entity3D - therefore a function from the Chem package is used here!</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Position: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Chem</span><span class="o">.</span><span class="n">get3DCoordinates</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">hasOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Orientation: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># create reader for input pharmacophores (format specified by file extension)</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreReader</span><span class="p">(</span><span class="s2">&quot;path_to_input_file.pml&quot;</span><span class="p">)</span>

<span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
<span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>

<span class="c1"># process pharmacophores one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">print_pharmacophore_properties</span><span class="p">(</span><span class="n">ph4</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="calculation-of-atom-properties">
<h1><span class="section-number">4. </span>Calculation of Atom Properties<a class="headerlink" href="#calculation-of-atom-properties" title="Link to this heading"></a></h1>
<p>The CDPL provides a wide panel of pre-defined atom and bond properties. All of these properties can be retrieved and set/calculated
by corresponding function calls as demonstrated in the following example scripts.</p>
<section id="calculation-of-atom-classification-properties">
<h2><span class="section-number">4.1. </span>Calculation of atom classification properties<a class="headerlink" href="#calculation-of-atom-classification-properties" title="Link to this heading"></a></h2>
<p>Atomic properties provide basic chemical information about each atom of a molecule. Such properties are
e.g. chemical element, formal charge, implicit hydrogen counts, hybridization states, aromaticity, and so on.
The following code snippet shows how to calculate and retrieve properties that provide higher-order information about
the atoms in a molecular graph.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The following examples use functionality provided by the <a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1Chem.html">CDPL.Chem</a> and
<a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1MolProp.html">CDPL.MolProp</a> package.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.MolProp</span> <span class="k">as</span> <span class="nn">MolProp</span>

<span class="k">def</span> <span class="nf">outputProperties</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs the corresponding properties of each atom of the provided molecular graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecular graph to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>
    <span class="n">MolProp</span><span class="o">.</span><span class="n">perceiveHBondDonorAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive H-bond donor atom types and set corresponding atom properties</span>
    <span class="n">MolProp</span><span class="o">.</span><span class="n">perceiveHBondAcceptorAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive H-bond acceptor atom types and set corresponding atom properties</span>

    <span class="n">hba_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">UNDEF</span>                   <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">NONE</span>                    <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_H2O</span>                   <span class="p">:</span> <span class="s1">&#39;O_H2O&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_UREA</span>                  <span class="p">:</span> <span class="s1">&#39;O_UREA&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_BARBITURIC_ACID</span>       <span class="p">:</span> <span class="s1">&#39;O_BARBITURIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_URIC_ACID</span>             <span class="p">:</span> <span class="s1">&#39;O_URIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ETHER</span>                 <span class="p">:</span> <span class="s1">&#39;O_ETHER&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_AMIDE</span>                 <span class="p">:</span> <span class="s1">&#39;O_AMIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_N_OXIDE</span>               <span class="p">:</span> <span class="s1">&#39;O_N_OXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ACID</span>                  <span class="p">:</span> <span class="s1">&#39;O_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ESTER</span>                 <span class="p">:</span> <span class="s1">&#39;O_ESTER&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_SULFOXIDE</span>             <span class="p">:</span> <span class="s1">&#39;O_SULFOXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_NITRO</span>                 <span class="p">:</span> <span class="s1">&#39;O_NITRO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_SELEN_OXIDE</span>           <span class="p">:</span> <span class="s1">&#39;O_SELEN_OXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ALDEHYD</span>               <span class="p">:</span> <span class="s1">&#39;O_ALDEHYD&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_KETONE</span>                <span class="p">:</span> <span class="s1">&#39;O_KETONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ALCOHOL</span>               <span class="p">:</span> <span class="s1">&#39;O_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_NH3</span>                   <span class="p">:</span> <span class="s1">&#39;N_NH3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_DIAMINE</span>               <span class="p">:</span> <span class="s1">&#39;N_DIAMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_MONO_DI_NITRO_ANILINE</span> <span class="p">:</span> <span class="s1">&#39;N_MONO_DI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_TRI_NITRO_ANILINE</span>     <span class="p">:</span> <span class="s1">&#39;N_TRI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_HALOGENO_ANILINE</span>      <span class="p">:</span> <span class="s1">&#39;N_HALOGENO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_ANILINE</span>               <span class="p">:</span> <span class="s1">&#39;N_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_NITRILE</span>               <span class="p">:</span> <span class="s1">&#39;N_NITRILE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZOLE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AZOLE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AMINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AMIDINE</span>               <span class="p">:</span> <span class="s1">&#39;N_AMIDINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZO</span>                   <span class="p">:</span> <span class="s1">&#39;N_AZO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AZINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_DIAZINE</span>               <span class="p">:</span> <span class="s1">&#39;N_DIAZINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_IMINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_IMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">S_SULFIDE</span>               <span class="p">:</span> <span class="s1">&#39;S_SULFIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">S_THIOUREA</span>              <span class="p">:</span> <span class="s1">&#39;S_THIOUREA&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">P_MONO_DI_PHOSPHINE</span>     <span class="p">:</span> <span class="s1">&#39;P_MONO_DI_PHOSPHINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">P_TRI_PHOSPHINE</span>         <span class="p">:</span> <span class="s1">&#39;P_TRI_PHOSPHINE&#39;</span> <span class="p">}</span>

    <span class="n">hbd_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">UNDEF</span>                       <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">NONE</span>                        <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">I_HI</span>                        <span class="p">:</span> <span class="s1">&#39;I_HI&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">BR_HBR</span>                      <span class="p">:</span> <span class="s1">&#39;BR_HBR&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">CL_HCL</span>                      <span class="p">:</span> <span class="s1">&#39;CL_HCL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_HSCN</span>                      <span class="p">:</span> <span class="s1">&#39;S_HSCN&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">F_HF</span>                        <span class="p">:</span> <span class="s1">&#39;F_HF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">H_H2</span>                        <span class="p">:</span> <span class="s1">&#39;H_H2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">C_HCN</span>                       <span class="p">:</span> <span class="s1">&#39;C_HCN&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">C_ETHINE</span>                    <span class="p">:</span> <span class="s1">&#39;C_ETHINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_HN3</span>                       <span class="p">:</span> <span class="s1">&#39;N_HN3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_NH3</span>                       <span class="p">:</span> <span class="s1">&#39;N_NH3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_NH4</span>                       <span class="p">:</span> <span class="s1">&#39;N_NH4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMINE</span>                     <span class="p">:</span> <span class="s1">&#39;N_AMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMINIUM</span>                   <span class="p">:</span> <span class="s1">&#39;N_AMINIUM&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_ANILINE</span>                   <span class="p">:</span> <span class="s1">&#39;N_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_MONO_DI_NITRO_ANILINE</span>     <span class="p">:</span> <span class="s1">&#39;N_MONO_DI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_TRI_NITRO_ANILINE</span>         <span class="p">:</span> <span class="s1">&#39;N_TRI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_PYRROLE</span>                   <span class="p">:</span> <span class="s1">&#39;N_PYRROLE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMIDE</span>                     <span class="p">:</span> <span class="s1">&#39;N_AMIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_IMINE</span>                     <span class="p">:</span> <span class="s1">&#39;N_IMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_IMINIUM</span>                   <span class="p">:</span> <span class="s1">&#39;N_IMINIUM&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_H2S</span>                       <span class="p">:</span> <span class="s1">&#39;S_H2S&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_HS</span>                        <span class="p">:</span> <span class="s1">&#39;S_HS&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_THIOL</span>                     <span class="p">:</span> <span class="s1">&#39;S_THIOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H3PO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H3PO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2CO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2CO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HCO3</span>                      <span class="p">:</span> <span class="s1">&#39;O_HCO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2O2</span>                      <span class="p">:</span> <span class="s1">&#39;O_H2O2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2O</span>                       <span class="p">:</span> <span class="s1">&#39;O_H2O&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CF3SO3H</span>                   <span class="p">:</span> <span class="s1">&#39;O_CF3SO3H&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HCLO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_HCLO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2SO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2SO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HNO3</span>                      <span class="p">:</span> <span class="s1">&#39;O_HNO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HSO4</span>                      <span class="p">:</span> <span class="s1">&#39;O_HSO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HNO2</span>                      <span class="p">:</span> <span class="s1">&#39;O_HNO2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_NH2OH</span>                     <span class="p">:</span> <span class="s1">&#39;O_NH2OH&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2PO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2PO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H3BO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H3BO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H4SIO4</span>                    <span class="p">:</span> <span class="s1">&#39;O_H4SIO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HPO4</span>                      <span class="p">:</span> <span class="s1">&#39;O_HPO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2BO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2BO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HO</span>                        <span class="p">:</span> <span class="s1">&#39;O_HO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_SULFONIC_ACID</span>             <span class="p">:</span> <span class="s1">&#39;O_SULFONIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_MONO_DI_NITRO_PHENOL</span>      <span class="p">:</span> <span class="s1">&#39;O_MONO_DI_NITRO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_ALCOHOL</span>          <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_ALCOHOL</span>                   <span class="p">:</span> <span class="s1">&#39;O_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_TRI_NITRO_PHENOL</span>          <span class="p">:</span> <span class="s1">&#39;O_TRI_NITRO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_PHENOL</span>           <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_PHENOL</span>                    <span class="p">:</span> <span class="s1">&#39;O_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CARBOXYLIC_ACID</span>           <span class="p">:</span> <span class="s1">&#39;O_CARBOXYLIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_CARBOXYCLIC_ACID</span> <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_CARBOXYCLIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_ENOL</span>                      <span class="p">:</span> <span class="s1">&#39;O_ENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_OXIME</span>                     <span class="p">:</span> <span class="s1">&#39;O_OXIME&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CL5_PHENOL</span>                <span class="p">:</span> <span class="s1">&#39;O_CL5_PHENOL&#39;</span> <span class="p">}</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom #</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">molgraph</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is std. hydrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isOrdinaryHydrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is heavy atom: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHeavy</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is unsaturated: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isUnsaturated</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is H-bond acceptor: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHBondAcceptor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">H-bond acceptor type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hba_type_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHBondAcceptorType</span><span class="p">(</span><span class="n">atom</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is H-bond donor: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHBondDonor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">H-bond donor type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hbd_type_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHBondDonorType</span><span class="p">(</span><span class="n">atom</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is carbonyl carbon: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isCarbonylLikeAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is amide carbon: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isAmideCenterAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is amide nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isAmideNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is invertible nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isInvertibleNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is planar nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isPlanarNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Usage: </span><span class="si">%s</span><span class="s1"> &lt;input mol. file&gt;&#39;</span> <span class="o">%</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">outputProperties</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="calculation-of-connectivity-properties">
<h2><span class="section-number">4.2. </span>Calculation of connectivity properties<a class="headerlink" href="#calculation-of-connectivity-properties" title="Link to this heading"></a></h2>
<p>The code snippet below shows to calculate various atom properties that depend on their connectivity to other atoms of
the molecular graph such as the number of connected carbon atoms, heteroatoms, halogens, heavy atoms, chain atoms,
ring atoms, aromatic atoms, and many more.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.MolProp</span> <span class="k">as</span> <span class="nn">MolProp</span>

<span class="k">def</span> <span class="nf">outputProperties</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs the corresponding properties of each atom of the provided molecular graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecular graph to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">vsepr_geom_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">UNDEF</span>                  <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">NONE</span>                   <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">LINEAR</span>                 <span class="p">:</span> <span class="s1">&#39;LINEAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_PLANAR</span>        <span class="p">:</span> <span class="s1">&#39;TRIGONAL_PLANAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TETRAHEDRAL</span>            <span class="p">:</span> <span class="s1">&#39;TETRAHEDRAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_BIPYRAMIDAL</span>   <span class="p">:</span> <span class="s1">&#39;TRIGONAL_BIPYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">OCTAHEDRAL</span>             <span class="p">:</span> <span class="s1">&#39;OCTAHEDRAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_BIPYRAMIDAL</span> <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_BIPYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_ANTIPRISMATIC</span>   <span class="p">:</span> <span class="s1">&#39;SQUARE_ANTIPRISMATIC&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">BENT</span>                   <span class="p">:</span> <span class="s1">&#39;BENT&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_PYRAMIDAL</span>     <span class="p">:</span> <span class="s1">&#39;TRIGONAL_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_PLANAR</span>          <span class="p">:</span> <span class="s1">&#39;SQUARE_PLANAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_PYRAMIDAL</span>       <span class="p">:</span> <span class="s1">&#39;SQUARE_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">T_SHAPED</span>               <span class="p">:</span> <span class="s1">&#39;T_SHAPED&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SEESAW</span>                 <span class="p">:</span> <span class="s1">&#39;SEESAW&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_PYRAMIDAL</span>   <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_PLANAR</span>      <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_PLANAR&#39;</span> <span class="p">}</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom #</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">molgraph</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected std. hydrogens (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getOrdinaryHydrogenCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected carbon atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">C</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected heteroatoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">HET</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected halogens: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected heavy atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHeavyAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected chain atoms (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitChainAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected chain atoms (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getChainAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected ring atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRingAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected aromatic atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getAromaticAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident single bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident single bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident double bonds: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">2</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident triple bonds: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">3</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident chain bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitChainBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident chain bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getChainBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident ring bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRingBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident aromatic bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getAromaticBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident heavy atom bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHeavyBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident rotatable bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRotatableBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Valence (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcExplicitValence</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Valence (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcValence</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Steric number: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcStericNumber</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">VSEPR coordination geometry: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">vsepr_geom_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getVSEPRCoordinationGeometry</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)])</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
       <span class="n">outputProperties</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
       <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="calculation-of-molecular-structure-descriptors">
<h1><span class="section-number">5. </span>Calculation of Molecular Structure Descriptors<a class="headerlink" href="#calculation-of-molecular-structure-descriptors" title="Link to this heading"></a></h1>
<p>The calculation of molecule structure descriptors is one of the fundamental operations in cheminformatics.
Such descriptors e.g. allow for a modeling and prediction of various structure-dependent properties
with mathematical methods. However, they can also be put to use for numerous other applications such as:</p>
<ul class="simple">
<li><p>Search for molecules that are structurally similar to a query molcule</p></li>
<li><p>Pre-filtering step for substructure searching in large chemical databases</p></li>
<li><p>Identifying the presence of particular functional groups or fragments in molecules</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1Descr.html">CDPL.Descr</a> package provides functionality for the generation
of various well-known molecule descriptors and fingerprints.</p>
</div>
<section id="extended-connectivity-fingerprints-ecfps">
<h2><span class="section-number">5.1. </span>Extended connectivity fingerprints (ECFPs)<a class="headerlink" href="#extended-connectivity-fingerprints-ecfps" title="Link to this heading"></a></h2>
<p>Morgan circular fingerprints, also known as extended connectivity fingerprints (ECFPs) <span id="id5">[<a class="reference internal" href="../bibliography.html#id11" title="David Rogers and Mathew Hahn. Extended-connectivity fingerprints. Journal of Chemical Information and Modeling, 50(5):742-754, 2010. PMID: 20426451. URL: https://doi.org/10.1021/ci100050t, arXiv:https://doi.org/10.1021/ci100050t, doi:10.1021/ci100050t.">3</a>]</span>,
are a type of structural fingerprint that encodes the local chemical environment of each atom in a
molecule as a particular bit in a bitset. They are widely used in cheminformatics for various tasks, such as similarity searching,
virtual screening, and machine learning.</p>
<p>The following code snippet calculates and outputs the ECFP fingerprints of the provided molecules.
The ECFP generation process can be influenced by the following parameters:</p>
<ul class="simple">
<li><p><em>num_bits</em>: The number of bits of the fingerprint (default: 1024)</p></li>
<li><p><em>radius</em>: Max. atom environment radius in number of bonds (default: 2)</p></li>
<li><p><em>inc_hs</em>: Whether to include explicit hydrogens (by default, the fingerprint is generated for the H-deplete molecular graph)</p></li>
<li><p><em>inc_config</em>: Whether to include atom chirality (by default, the fingerprint is generated for the H-deplete molecular graph)</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Descr</span> <span class="k">as</span> <span class="nn">Descr</span>
<span class="kn">import</span> <span class="nn">CDPL.Util</span> <span class="k">as</span> <span class="nn">Util</span>

<span class="k">def</span> <span class="nf">genECFP</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">num_bits</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">radius</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">inc_hs</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">inc_config</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Util</span><span class="o">.</span><span class="n">BitSet</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the binary ECFP for the given molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to process.</span>
<span class="sd">    - num_bits (int): Number of bits of the fingerprint.</span>
<span class="sd">    - radius (int): Max. atom environment radius in number of bonds.</span>
<span class="sd">    - inc_hs (bool): Whether to include explicit hydrogens.</span>
<span class="sd">    - inc_config (bool): Whether to include atom chirality.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Util.BitSet: The generated fingerprint.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># calculate implicit hydrogen counts (if not yet done)</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                      <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                      <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>               <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">ecfp_gen</span> <span class="o">=</span> <span class="n">Descr</span><span class="o">.</span><span class="n">CircularFingerprintGenerator</span><span class="p">()</span>    <span class="c1"># create ECFP generator instance</span>

    <span class="k">if</span> <span class="n">inc_config</span><span class="p">:</span>
        <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">includeChirality</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>                <span class="c1"># allow atom chirality to have an impact on the ECFP generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcAtomStereoDescriptors</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>     <span class="c1"># calculate atom stereo descriptors and set corresponding property for all atoms</span>

    <span class="k">if</span> <span class="n">inc_hs</span><span class="p">:</span>
        <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">includeHydrogens</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>                <span class="c1"># include explicit hydrogens in the ECFP generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                 <span class="c1"># make any implicit hydrogens explicit</span>

    <span class="n">fp</span> <span class="o">=</span> <span class="n">Util</span><span class="o">.</span><span class="n">BitSet</span><span class="p">()</span>                                 <span class="c1"># create fingerprint bitset</span>
    <span class="n">fp</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="n">num_bits</span><span class="p">)</span>                                <span class="c1"># set desired fingerprint size</span>

    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">setNumIterations</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span>                  <span class="c1"># set num. iterations (=atom. env. radius)</span>
    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                             <span class="c1"># extract chracteristic structural features</span>
    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">setFeatureBits</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span>                        <span class="c1"># set bits associated with the extracted structural features</span>

    <span class="c1"># if needed, fp could be converted into a numpy single precision float array as follows:</span>
    <span class="c1"># fp = numpy.array(fp, dtype=numpy.float32)</span>

    <span class="k">return</span> <span class="n">fp</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># Settings</span>
<span class="n">num_bits</span> <span class="o">=</span> <span class="mi">1024</span>   <span class="c1"># fingerprint size</span>
<span class="n">radius</span> <span class="o">=</span> <span class="mi">4</span>        <span class="c1"># atom environment radius</span>
<span class="n">inc_hs</span> <span class="o">=</span> <span class="kc">True</span>     <span class="c1"># include explicit hydrogens</span>
<span class="n">inc_config</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># include atom chirality</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">fp</span> <span class="o">=</span> <span class="n">genECFP</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">num_bits</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">inc_hs</span><span class="p">,</span> <span class="n">inc_config</span><span class="p">)</span>

        <span class="c1"># do something useful with the fingerprint</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="fame-atom-environment-fingerprints">
<h2><span class="section-number">5.2. </span>FAME atom environment fingerprints<a class="headerlink" href="#fame-atom-environment-fingerprints" title="Link to this heading"></a></h2>
<p>This type of fingerprint encodes the local environment of individual atoms up to a configurable maximum bond-path length.
The descriptor was developed for the classification of atoms in the well-known site of metabolism prediction
software <em>FAME</em> <span id="id6">[<a class="reference internal" href="../bibliography.html#id5" title="Martin Šícho, Conrad Stork, Angelica Mazzolari, Christina de Bruyn Kops, Alessandro Pedretti, Bernard Testa, Giulio Vistoli, Daniel Svozil, and Johannes Kirchmair. Fame 3: predicting the sites of metabolism in synthetic compounds and natural products for phase 1 and phase 2 metabolic enzymes. Journal of Chemical Information and Modeling, 59(8):3400-3412, 2019. PMID: 31361490. URL: https://doi.org/10.1021/acs.jcim.9b00376, arXiv:https://doi.org/10.1021/acs.jcim.9b00376, doi:10.1021/acs.jcim.9b00376.">16</a>]</span>.</p>
<p>The following code snippet calculates and outputs the FAME descriptor for each atom of the provided molecules with
the parameter <em>radius</em> specifying the max. atom environment radius in number of bonds (default: 2).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">genFAMEDescriptor</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">,</span> <span class="n">radius</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the FAME descriptor for the given atom of the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ctr_atom (Chem.Atom): Atom for which the FAME descriptor is to be calculated.</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecule to process.</span>
<span class="sd">    - radius (int): Max. atom environment radius in number of bonds.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - numpy.array: The generated FAME descriptor.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">env</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span>                                                      <span class="c1"># for storing of extracted environment atoms</span>
    <span class="n">descr</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">Chem</span><span class="o">.</span><span class="n">SybylAtomType</span><span class="o">.</span><span class="n">MAX_TYPE</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">radius</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>

    <span class="n">Chem</span><span class="o">.</span><span class="n">getEnvironment</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">env</span><span class="p">)</span>                       <span class="c1"># extract environment of center atom reaching</span>
                                                                               <span class="c1"># out up to &#39;radius&#39; bonds</span>
    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">env</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>                                                     <span class="c1"># iterate over extracted environment atoms</span>
        <span class="n">sybyl_type</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getSybylType</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>                                   <span class="c1"># retrieve Sybyl type of environment atom</span>
        <span class="n">top_dist</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getTopologicalDistance</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">,</span> <span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)</span>       <span class="c1"># get top. distance between center atom and environment atom</span>
        <span class="n">descr</span><span class="p">[</span><span class="n">top_dist</span> <span class="o">*</span> <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">SybylAtomType</span><span class="o">.</span><span class="n">MAX_TYPE</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">sybyl_type</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># instead of 1 (= Sybyl type presence) also any other numeric atom</span>
                                                                               <span class="c1"># property could be summed up here</span>
    <span class="k">return</span> <span class="n">descr</span>

<span class="k">def</span> <span class="nf">procMolecule</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecule to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>     <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>    <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                   <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                   <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>            <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSybylAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive Sybyl atom types and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcTopologicalDistanceMatrix</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate topological distance matrix and store as Chem.MolecularGraph property</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="n">descr</span> <span class="o">=</span> <span class="n">genFAMEDescriptor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>     <span class="c1"># generate atom environment descriptor using a radius of five bonds</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">descr</span><span class="p">)</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># process molecules one after the other</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">procMolecule</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="force-field-calculations">
<h1><span class="section-number">6. </span>Force Field Calculations<a class="headerlink" href="#force-field-calculations" title="Link to this heading"></a></h1>
<p>Force fields are computational methods that estimate the forces between atoms within molecules and also between molecules.
Since they allow for a computationally inexpensive, but nevertheless relatively accurate calculation of the potential energy of a system of atoms,
they are essential for many tasks in the field of molecular modeling such as conformer generation and molecular dynamics (MD) simulations.
One of the widely recognized force fields is the <em>Merck Molecular Force Field</em> (MMFF94)
<span id="id7">[<a class="reference internal" href="../bibliography.html#id10" title="Thomas A. Halgren. Merck molecular force field. i. basis, form, scope, parameterization, and performance of mmff94. Journal of Computational Chemistry, 17(5-6):490-519, 1996. URL: https://onlinelibrary.wiley.com/doi/abs/10.1002/%28SICI%291096-987X%28199604%2917%3A5/6%3C490%3A%3AAID-JCC1%3E3.0.CO%3B2-P, arXiv:https://onlinelibrary.wiley.com/doi/pdf/10.1002/%28SICI%291096-987X%28199604%2917%3A5/6%3C490%3A%3AAID-JCC1%3E3.0.CO%3B2-P, doi:https://doi.org/10.1002/(SICI)1096-987X(199604)17:5/6&lt;490::AID-JCC1&gt;3.0.CO;2-P.">6</a>]</span> which has been designed to be applicable to a
broad range of organic molecules and thus found widespread use.</p>
<section id="calculation-of-mmff94-atom-charges">
<h2><span class="section-number">6.1. </span>Calculation of MMFF94 atom charges<a class="headerlink" href="#calculation-of-mmff94-atom-charges" title="Link to this heading"></a></h2>
<p>The MMFF94 force field provides a method and associated parameter set for the calculation of partial atomic charges.
They are not only crucial for the estimation of intra- and intermolecular electrostatic interactions but can also be
used, e.g., as atom descriptors for ML-based model building.</p>
<p>The following code snippet calculates and outputs the MMFF94 charges of the atoms for a given list of molecules:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Force field related functionality is provided via the <a class="reference external" href="../cdpl_api_doc/python_api_doc/namespaceCDPL_1_1ForceField.html">CDPL.ForceField</a> package.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ForceField</span> <span class="k">as</span> <span class="nn">ForceField</span>

 <span class="k">def</span> <span class="nf">calc_and_output_charges</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and outputs the MMFF94 charges of the atoms for the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        mol (Chem.BasicMolecule): The molecule for which the MMFF94 charges are to be calculated.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Various preprocessing steps to prepare the molecule</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>               <span class="c1"># make all implicit hydrogens explicit</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">ForceField</span><span class="o">.</span><span class="n">perceiveMMFF94AromaticRings</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># perceive aromatic rings according to the MMFF94 aroamticity model and store data as Chem.MolecularGraph property</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">assignMMFF94AtomTypes</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># perceive MMFF94 atom types (tolerant mode) set corresponding property for all atoms</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">assignMMFF94BondTypeIndices</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive MMFF94 bond types (tolerant mode) set corresponding property for all bonds</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">calcMMFF94AtomCharges</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># calculate MMFF94 atom charges (tolerant mode) set corresponding property for all atoms</span>


    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- MMFF94 partial charges&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Atom #</span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">getIndex</span><span class="p">()),</span> <span class="nb">str</span><span class="p">(</span><span class="n">ForceField</span><span class="o">.</span><span class="n">getMMFF94Charge</span><span class="p">(</span><span class="n">atom</span><span class="p">))))</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_of_BasicMolecules</span><span class="p">]</span>

<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="n">calc_and_output_charges</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="substructure-searching-and-matching">
<h1><span class="section-number">7. </span>Substructure Searching and Matching<a class="headerlink" href="#substructure-searching-and-matching" title="Link to this heading"></a></h1>
<p>Substructure searching (a subgraph isomorphism problem) is one of the most fundamental operations in cheminformatics and allows for the
identification of molecules that contain a specific structural motif or pattern.
Substructure searching needs to be performed in various application areas, such as:</p>
<ul class="simple">
<li><p><em>Drug Discovery</em>: Identifying molecules that contain a particular pharmacophore or motif</p></li>
<li><p><em>Chemical Database Querying</em>: Filtering large chemical databases to retrieve molecules of interest</p></li>
<li><p><em>Reactivity Analysis</em>: Identifying the presence of particular functional groups or fragments in molecules</p></li>
</ul>
<section id="filtering-molecules-based-on-a-smarts-pattern">
<h2><span class="section-number">7.1. </span>Filtering molecules based on a SMARTS pattern<a class="headerlink" href="#filtering-molecules-based-on-a-smarts-pattern" title="Link to this heading"></a></h2>
<p>The SMARTS notation (SMiles ARbitrary Target Specification) is a language used to describe structural patterns. It’s an
extension of the SMILES notation and allows for more complex and specific pattern descriptions.
The CDPL provides substructure searching functionality via the class <a class="reference external" href="../cdpl_api_doc/python_api_doc/classCDPL_1_1Chem_1_1SubstructureSearch.html">Chem.SubstructureSearch</a>.
The code snippet below demonstrates how this class can be used to identify molecules that match a specific structural motif described by a SMARTS pattern.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">filterMoleculesBySmarts</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">smarts_pattern</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">quiet</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filters molecules from the input file that match the provided SMARTS pattern and writes them to the output file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        input_file (str): Path to the input file containing the molecules to be filtered.</span>
<span class="sd">        output_file (str): Path to the output file to which the filtered molecules are written.</span>
<span class="sd">        smarts_pattern (str): SMARTS pattern describing the structural motif to be matched.</span>
<span class="sd">        quiet (bool): If set to True, no progress information is printed to the console.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sub_srch_ptn</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">parseSMARTS</span><span class="p">(</span><span class="n">smarts_pattern</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">initSubstructureSearchQuery</span><span class="p">(</span><span class="n">sub_srch_ptn</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error: parsing of SMARTS pattern failed: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="n">substr_srch</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SubstructureSearch</span><span class="p">(</span><span class="n">sub_srch_ptn</span><span class="p">)</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
    <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>
    <span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">or</span> <span class="sa">f</span><span class="s1">&#39;Molecule_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>

            <span class="n">Chem</span><span class="o">.</span><span class="n">initSubstructureSearchTarget</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">substr_srch</span><span class="o">.</span><span class="n">mappingExists</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">quiet</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39; -&gt; substructure found, forwarding molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1"> to output file&#39;</span><span class="p">)</span>
                <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="n">quiet</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39; -&gt; substructure not found in molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">filterMoleculesBySmarts</span><span class="p">(</span><span class="s2">&quot;input.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;output.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;[#6]1:[#6]:[#6]:[#6]:[#6]:[#6]:1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="CDPL Python Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../cdpl_python_cookbook/index.html" class="btn btn-neutral float-right" title="CDPL Python Cookbook" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Thomas Seidel and Oliver Wieder.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>