<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>1. Molecular input/output &mdash; CDPKit 1.0.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/tabs.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/tabs.js"></script>
        <script src="../_static/docs-versions-menu.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="CDPL Python Cookbook" href="../cdpl_python_cookbook/index.html" />
    <link rel="prev" title="CDPL Python Tutorial" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #aaaaaa" >
            <a href="../index.html">
            <img src="../_static/Logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">CDPL Python Tutorial</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">1. Molecular input/output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#reading-molecules-from-an-input-file">1.1. Reading Molecules from an Input File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#retrieving-mdl-structure-data-from-sd-files">1.2. Retrieving MDL Structure Data from SD Files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#writing-molecules-to-an-output-file">1.3. Writing Molecules to an Output File</a></li>
<li class="toctree-l3"><a class="reference internal" href="#extracting-atom-environments-from-molecules-and-extract-smiles-strings-for-the-environments">1.4. Extracting Atom Environments from Molecules and extract SMILES strings for the environments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#chembl-molecule-standardization-and-parent-structure-extraction">1.5. ChEMBL Molecule Standardization and Parent Structure Extraction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#protonation-of-functional-groups">1.6. Protonation of Functional Groups</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#molecular-conformation-and-generation">2. Molecular Conformation and Generation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generating-low-energy-3d-conformations">2.1. Generating low-energy 3D Conformations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generating-conformation-ensembles">2.2. Generating Conformation Ensembles</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#pharmacophore-generation-and-processing">3. Pharmacophore Generation and Processing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#ligand-based-pharmacophore-generation-using-conformations">3.1. Ligand-based Pharmacophore Generation using Conformations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#d-ligand-based-pharmacophore-generation-for-a-set-of-conformations">3.2. 3D Ligand-based Pharmacophore Generation for a set of Conformations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#aligne-ligand-based-pharmacophores-for-a-set-of-conformations">3.3. Aligne Ligand-based Pharmacophores for a set of Conformations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pharmacophore-features-and-feature-types">3.4. Pharmacophore Features and Feature Types</a></li>
<li class="toctree-l3"><a class="reference internal" href="#d-structure-based-pharmacophore-generation">3.5. 3D Structure-based Pharmacophore Generation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculation-of-atom-and-bond-properties">4. Calculation of Atom and Bond Properties</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calculation-of-atomic-properties">4.1. Calculation of Atomic Properties</a></li>
<li class="toctree-l3"><a class="reference internal" href="#calculation-of-connectivity-properties">4.2. Calculation of Connectivity Properties</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#calculation-of-molecule-and-pharmacophore-descriptors">5. Calculation of Molecule and Pharmacophore Descriptors</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extended-connectivity-fingerprints-ecfps">5.1. Extended Connectivity Fingerprints (ECFPs)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fame-atom-environment-fingerprints">5.2. FAME Atom Environment Fingerprints</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#force-field-calculations">6. Force Field Calculations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#mmff94-atom-charges-calculation">6.1. MMFF94 Atom Charges Calculation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#substructure-searching-and-matching">7. Substructure Searching and Matching</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#filtering-molecules-based-on-a-smarts-pattern">7.1. Filtering Molecules Based on a SMARTS Pattern</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cdpl_python_cookbook/index.html">CDPL Python Cookbook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cdpl_api_doc.html">CDPL API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="../genindex.html">Index</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #aaaaaa" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">CDPKit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="index.html">CDPL Python Tutorial</a> &raquo;</li>
      <li><span class="section-number">1. </span>Molecular input/output</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/cdpl_python_tutorial/cdpl_python_tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="molecular-input-output">
<h1><span class="section-number">1. </span>Molecular input/output<a class="headerlink" href="#molecular-input-output" title="Permalink to this heading"></a></h1>
<section id="reading-molecules-from-an-input-file">
<h2><span class="section-number">1.1. </span>Reading Molecules from an Input File<a class="headerlink" href="#reading-molecules-from-an-input-file" title="Permalink to this heading"></a></h2>
<p>To kick off your cheminformatics journey, you’ll frequently need to read or write molecular data from a variety of file formats.
Using CDPKit, you can easily read and write molecules from various file formats, including <em>SDF</em>, <em>SMILES</em>, <em>MOL</em>, <em>MOL2</em>, <em>CDF</em>, and others.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The CDF format is a binary format that can be used to store molecules and their associated metadata and is CDPKit’s native file format.</p>
</div>
<p>The <a class="reference external" href="../python_api_doc/classCDPL_1_1Chem_1_1MoleculeReader.html">Chem.MoleculeReader</a> is versatile and can interpret different types of formats (SDF, SMILES, MOL, MOL2, CDF, etc.),
eliminating the need to call a specific reader for each format.
Here is a simple example of how to read molecules from an input file:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The module <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1Chem.html">CDPL.Chem</a> provides functionality for the processing of chemical data.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

 <span class="k">def</span> <span class="nf">processMoleculesFromFile</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
     <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Reads molecules from the provided input file and outputs the number of atoms and bonds for each molecule.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - input_file (str): Path to the input file.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># Create reader for input molecules (format specified by file extension)</span>
     <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>

     <span class="c1"># create an instance of the default implementation of the Chem.Molecule interface</span>
     <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>

     <span class="c1"># Iterate over each molecule and print atom and bond count</span>
     <span class="k">try</span><span class="p">:</span>
         <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
             <span class="k">try</span><span class="p">:</span>
                 <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Molecule with&#39;</span><span class="p">,</span> <span class="n">mol</span><span class="o">.</span><span class="n">numAtoms</span><span class="p">,</span> <span class="s1">&#39;atoms and&#39;</span><span class="p">,</span> <span class="n">mol</span><span class="o">.</span><span class="n">numBonds</span><span class="p">,</span> <span class="s1">&#39;bonds&#39;</span><span class="p">)</span>
             <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>
 <span class="n">processMoleculesFromFile</span><span class="p">(</span><span class="s2">&quot;path_to_mol_file.mol2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="retrieving-mdl-structure-data-from-sd-files">
<h2><span class="section-number">1.2. </span>Retrieving MDL Structure Data from SD Files<a class="headerlink" href="#retrieving-mdl-structure-data-from-sd-files" title="Permalink to this heading"></a></h2>
<p>Sometimes, one might want to read the metadata associated with a molecule in an SD file. CDPKit provides a convenient way to do this shown
in the following example:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can either use Chem.MoleculeReader to read the structure data of each molecule in an SD file or
use the specific reader like e.g. Chem.FileSDFMoleculeReader to read the structure data of each molecule in an SD file.
The convenient way is to use Chem.MoleculeReader.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">processStructureDataFromSD</span><span class="p">(</span><span class="n">input_sd_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd"> Retrieves the structure data of each molecule in the provided SD file and outputs it to the console.</span>

<span class="sd"> Parameters:</span>
<span class="sd"> - input_sd_file (str): Path to the input SD file.</span>
<span class="sd"> &quot;&quot;&quot;</span>
     <span class="c1"># Create reader for MDL SD-files</span>
     <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">FileSDFMoleculeReader</span><span class="p">(</span><span class="n">input_sd_file</span><span class="p">)</span>

     <span class="c1"># create an instance of the default implementation of the Chem.Molecule interface</span>
     <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>

     <span class="c1"># Iterate over each molecule in the file and retrieve structure data</span>
     <span class="k">try</span><span class="p">:</span>
         <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
             <span class="k">try</span><span class="p">:</span>
                 <span class="k">if</span> <span class="ow">not</span> <span class="n">Chem</span><span class="o">.</span><span class="n">hasStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Error: no structure data available for molecule&#39;</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">))</span>
                     <span class="k">continue</span>

                 <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span> <span class="c1"># retrieve structure data</span>

                 <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Structure data (</span><span class="si">%s</span><span class="s1"> entries) of molecule </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1">:</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_data</span><span class="p">)),</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)))</span>

                 <span class="k">for</span> <span class="n">entry</span> <span class="ow">in</span> <span class="n">struct_data</span><span class="p">:</span> <span class="c1"># iterate of structure data entries consisting of a header line and the actual data</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Header:&#39;</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">header</span><span class="p">)</span>
                     <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data:&#39;</span><span class="p">,</span> <span class="n">entry</span><span class="o">.</span><span class="n">data</span><span class="p">)</span>
             <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                 <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
         <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

 <span class="c1">##########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>
 <span class="n">processStructureDataFromSD</span><span class="p">(</span><span class="s2">&quot;path_to_sd_file.sdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="writing-molecules-to-an-output-file">
<h2><span class="section-number">1.3. </span>Writing Molecules to an Output File<a class="headerlink" href="#writing-molecules-to-an-output-file" title="Permalink to this heading"></a></h2>
<p>Once you’ve processed or analyzed your molecules, you may want to save them to an output file.
The format of the output file is determined by its file extension, making it easy to save your molecules in formats such as SDF, MOL, MOL2, CDF, and others.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Ensure that the file extension you provide matches the desired output format. For instance, use <cite>.sdf</cite> for Structure-Data Files, <cite>.mol</cite> for MDL Molfiles, and so on.</p>
</div>
<p>Here’s a simple example of how to write a list of molecules to an output file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">molsToFiles</span><span class="p">(</span><span class="n">mols</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">],</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Writes a list of molecules to the specified output file.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mols (list): List of CDPKit molecules to write to the output file.</span>
<span class="sd">    - output_file (str): Path to the output file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create a writer for the output molecules (format specified by file extension)</span>
    <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>
 <span class="n">molsToFiles</span><span class="p">(</span><span class="n">mols</span><span class="p">,</span> <span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="extracting-atom-environments-from-molecules-and-extract-smiles-strings-for-the-environments">
<h2><span class="section-number">1.4. </span>Extracting Atom Environments from Molecules and extract SMILES strings for the environments<a class="headerlink" href="#extracting-atom-environments-from-molecules-and-extract-smiles-strings-for-the-environments" title="Permalink to this heading"></a></h2>
<p>To understand the local chemical environments around specific atoms in a molecule, we can extract the structural environments of atoms
within a given molecular graph and output them as SMILES strings.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">printAtomEnv</span><span class="p">(</span><span class="n">mols</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Extracts the atom environments of each atom in the provided list of molecules and outputs them as SMILES strings.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mols (list): List of CDPKit molecules to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

        <span class="n">frag</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span>                       <span class="c1"># for storing extracted atom environments</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom environments (radius = 3 bonds)&#39;</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">getEnvironment</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">mol</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">frag</span><span class="p">)</span>     <span class="c1"># extract environment of atom reaching out up to three bonds</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveComponents</span><span class="p">(</span><span class="n">frag</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># perceive molecular graph components (required for SMILES generation)</span>

            <span class="n">smiles</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">generateSMILES</span><span class="p">(</span><span class="n">frag</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># generate non-canonical SMILES string with explicit hydrogen atoms</span>

            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Atom #</span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">mol</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)),</span> <span class="n">smiles</span><span class="p">))</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span> <span class="c1"># Example list of BasicMolecules</span>
<span class="n">printAtomEnv</span><span class="p">(</span><span class="n">mols</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="chembl-molecule-standardization-and-parent-structure-extraction">
<h2><span class="section-number">1.5. </span>ChEMBL Molecule Standardization and Parent Structure Extraction<a class="headerlink" href="#chembl-molecule-standardization-and-parent-structure-extraction" title="Permalink to this heading"></a></h2>
<p>CDPKit provides a convenient way to standardize molecules using the ChEMBL standardization pipeline.
This process can be used to ensure that molecules are represented in a consistent and standardized manner, which is crucial for many cheminformatics tasks.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">standardize</span><span class="p">(</span><span class="n">chembl_proc</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="p">,</span> <span class="n">in_mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Performs ChEMBL molecule standardization and parent structure extraction (optional) for a given input molecule using a provided Chem.ChEMBLStandardizer instance.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - chembl_proc (Chem.ChEMBLStandardizer): Instance of the Chem.ChEMBLStandardizer class.</span>
<span class="sd">    - in_mol (Chem.Molecule): Input molecule to standardize.</span>
<span class="sd">    - out_mol (Chem.Molecule): Output molecule to store the standardized molecule.</span>
<span class="sd">    - proc_excluded (bool): If True, molecules flagged as excluded will be processed.</span>
<span class="sd">    - extract_parent (bool): If True, the parent structure will be extracted.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Chem.ChEMBLStandardizer.ChangeFlags: Flags indicating the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># here, the standardization is carried out on a copy of the read input molecule</span>
    <span class="c1"># (if only one molecule instance gets provided as argument, modifications will be made in-place)</span>
    <span class="n">change_flags</span> <span class="o">=</span> <span class="n">chembl_proc</span><span class="o">.</span><span class="n">standardize</span><span class="p">(</span><span class="n">in_mol</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="c1"># perform parent structure extraction (optional)</span>
        <span class="n">change_flags</span> <span class="o">&amp;=</span> <span class="o">~</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span>  <span class="c1"># clear excluded flag possibly set by the standardization</span>
                                                       <span class="c1"># procedure (might change after salt stripping)</span>
        <span class="n">change_flags</span> <span class="o">|=</span> <span class="n">chembl_proc</span><span class="o">.</span><span class="n">getParent</span><span class="p">(</span><span class="n">out_mol</span><span class="p">)</span>     <span class="c1"># extract parent structure (in-place) and add information</span>
                                                       <span class="c1"># about the carried out modifcations</span>
    <span class="k">return</span> <span class="n">change_flags</span>

<span class="k">def</span> <span class="nf">getListOfChangesString</span><span class="p">(</span><span class="n">change_flags</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a string listing the carried out modifications.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - change_flags (Chem.ChEMBLStandardizer.ChangeFlags): Flags indicating the carried out modifications.</span>
<span class="sd">    - verbose (bool): If True, the string will contain a detailed list of the carried out modifications.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: String listing the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">changes</span> <span class="o">=</span> <span class="s1">&#39;   Carried out modifications:&#39;</span>

    <span class="c1"># List of possible changes</span>
    <span class="n">change_list</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXPLICIT_HYDROGENS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Explicit hydrogens removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">UNKNOWN_STEREO_STANDARDIZED</span><span class="p">,</span> <span class="s1">&#39;Undefined stereocenter information standardized&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">BONDS_KEKULIZED</span><span class="p">,</span> <span class="s1">&#39;Kekule structure generated&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">STRUCTURE_NORMALIZED</span><span class="p">,</span> <span class="s1">&#39;Functional groups normalized&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">CHARGES_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Number of charged atoms reduced&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">TARTRATE_STEREO_CLEARED</span><span class="p">,</span> <span class="s1">&#39;Configuration of chiral tartrate atoms set to undefined&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">STRUCTURE_2D_CORRECTED</span><span class="p">,</span> <span class="s1">&#39;2D structure corrected&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ISOTOPE_INFO_CLEARED</span><span class="p">,</span> <span class="s1">&#39;Isotope information cleared&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">SALT_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Salt components removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">SOLVENT_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Solvent components removed&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">DUPLICATE_COMPONENTS_REMOVED</span><span class="p">,</span> <span class="s1">&#39;Duplicate components removed&#39;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">flag</span><span class="p">,</span> <span class="n">description</span> <span class="ow">in</span> <span class="n">change_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">flag</span><span class="p">:</span>
            <span class="n">changes</span> <span class="o">+=</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">    * &#39;</span> <span class="o">+</span> <span class="n">description</span>

    <span class="k">return</span> <span class="n">changes</span>

<span class="k">def</span> <span class="nf">getLogMessage</span><span class="p">(</span><span class="n">change_flags</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">ChangeFlags</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">extract_parent</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">mol_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">verbose</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a log message describing the carried out modifications.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - change_flags (Chem.ChEMBLStandardizer.ChangeFlags): Flags indicating the carried out modifications.</span>
<span class="sd">    - proc_excluded (bool): If True, molecules flagged as excluded will be processed.</span>
<span class="sd">    - extract_parent (bool): If True, the parent structure will be extracted.</span>
<span class="sd">    - mol_id (str): Identifier of the molecule.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: Log message describing the carried out modifications.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span><span class="p">)</span> <span class="ow">and</span> <span class="n">proc_excluded</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: discarded (flagged as excluded)&#39;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">proc_excluded</span> <span class="ow">and</span> <span class="p">(</span><span class="n">change_flags</span> <span class="o">&amp;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="o">.</span><span class="n">EXCLUDED</span><span class="p">):</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: forwarded unchanged (flagged as excluded)&#39;</span>

    <span class="k">if</span> <span class="n">change_flags</span><span class="p">:</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: modified</span><span class="se">\n</span><span class="si">{</span><span class="n">getListOfChangesString</span><span class="p">(</span><span class="n">change_flags</span><span class="p">,</span> <span class="n">verbose</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">: forwarded unchanged&#39;</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="n">chembl_proc</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ChEMBLStandardizer</span><span class="p">()</span>

<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="n">in_mol</span> <span class="o">=</span> <span class="n">mol</span>
    <span class="n">out_mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>
    <span class="n">change_flags</span> <span class="o">=</span> <span class="n">standardize</span><span class="p">(</span><span class="n">chembl_proc</span><span class="p">,</span> <span class="n">in_mol</span><span class="p">,</span> <span class="n">out_mol</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">extract_parent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">in_mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">or</span> <span class="sa">f</span><span class="s1">&#39;Molecule_</span><span class="si">{</span><span class="n">mols</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
    <span class="n">log_msg</span> <span class="o">=</span> <span class="n">getLogMessage</span><span class="p">(</span><span class="n">change_flags</span><span class="p">,</span> <span class="n">proc_excluded</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">extract_parent</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mol_id</span><span class="o">=</span><span class="n">mol_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">log_msg</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="protonation-of-functional-groups">
<h2><span class="section-number">1.6. </span>Protonation of Functional Groups<a class="headerlink" href="#protonation-of-functional-groups" title="Permalink to this heading"></a></h2>
<p>In the realm of computational chemistry and molecular modeling, understanding and predicting the behavior of molecules often hinges on the finer details.
One such critical detail is the protonation state of a molecule.
Protonation states influence a molecule’s charge, conformation, and reactivity, playing a vital role in processes like drug binding,
enzymatic reactions, and more.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The method <a class="reference external" href="../python_api_doc/classCDPL_1_1Chem_1_1ProtonationStateStandardizer.html">Chem.ProtonationStateStandardizer.standardize</a>
implements the protonation state generation algorithm.</p>
</div>
<p>In the following code snippet we will show how to protonate/deprotonate functional groups of a molecule for a given pH value and ionic strength.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create and initialize an instance of the class Chem.ProtonationStateStandardizer which</span>
<span class="c1"># implements the protonation state generation algorithm</span>
<span class="n">prot_state_gen</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ProtonationStateStandardizer</span><span class="p">()</span>

<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="c1"># compose a simple molecule identifier</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># protonate/deprotonate functional groups for phys. conditions</span>
            <span class="n">prot_state_gen</span><span class="o">.</span><span class="n">standardize</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">ProtonationStateStandardizer</span><span class="o">.</span><span class="n">PHYSIOLOGICAL_CONDITION_STATE</span><span class="p">)</span>

            <span class="c1"># enforce an update of the molecule components list (structure might have changed)</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveComponents</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing or output of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="molecular-conformation-and-generation">
<h1><span class="section-number">2. </span>Molecular Conformation and Generation<a class="headerlink" href="#molecular-conformation-and-generation" title="Permalink to this heading"></a></h1>
<p>Molecular conformation refers to the spatial arrangement of atoms in a molecule. Each distinct spatial arrangement is a unique conformation.
In the realm of cheminformatics and molecular modeling, the ability to generate and analyze different conformations of a molecule is crucial.
This is especially important in drug design, where the biological activity of a molecule can be highly dependent on its conformation.</p>
<p>One of the notable conformation generation tools included in the CDPKit is <cite>CONFORGE</cite>. For details on its implementation, uses, and performance
see <span id="id1">[<a class="reference internal" href="../bibliography.html#id5" title="Thomas Seidel, Christian Permann, Oliver Wieder, Stefan M. Kohlbacher, and Thierry Langer. High-quality conformer generation with conforge: algorithm and performance assessment. Journal of Chemical Information and Modeling, 0(0):null, 0. PMID: 37624145. URL: https://doi.org/10.1021/acs.jcim.3c00563, arXiv:https://doi.org/10.1021/acs.jcim.3c00563, doi:10.1021/acs.jcim.3c00563.">5</a>]</span>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The module CDPL.ConfGen provides functionality for the generation of conformation ensembles in own scripts.</p>
</div>
<section id="generating-low-energy-3d-conformations">
<h2><span class="section-number">2.1. </span>Generating low-energy 3D Conformations<a class="headerlink" href="#generating-low-energy-3d-conformations" title="Permalink to this heading"></a></h2>
<p>This process involves generating 3D structures for molecules that may initially be represented in a 2D format.</p>
<p>Here’s how you can generate low-energy 3D conformations for a molecule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ConfGen</span> <span class="k">as</span> <span class="nn">ConfGen</span>

 <span class="k">def</span> <span class="nf">generate3dConformation</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">struct_gen</span><span class="p">:</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">StructureGenerator</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
     <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Generates a low-energy 3D structure of the argument molecule using the provided initialized ConfGen.StructureGenerator instance.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - mol (Chem.Molecule): Molecule to generate a 3D structure for.</span>
<span class="sd">     - struct_gen (ConfGen.StructureGenerator): Instance of the ConfGen.StructureGenerator class.</span>

<span class="sd">     Returns:</span>
<span class="sd">     - int: Status code indicating the success of the 3D structure generation.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># prepare the molecule for 3D structure generation</span>
     <span class="n">ConfGen</span><span class="o">.</span><span class="n">prepareForConformerGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># generate the 3D structure</span>
     <span class="n">status</span> <span class="o">=</span> <span class="n">struct_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># if sucessful, store the generated conformer ensemble as</span>
     <span class="c1"># per atom 3D coordinates arrays (= the way conformers are represented in CDPKit)</span>
     <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span><span class="p">:</span>
         <span class="n">struct_gen</span><span class="o">.</span><span class="n">setCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># return status code</span>
     <span class="k">return</span> <span class="n">status</span>

 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="c1">#Flags</span>
 <span class="n">max_time</span> <span class="o">=</span> <span class="mi">3600</span> <span class="c1"># Max. allowed molecule processing time in seconds (default: 3600 sec)</span>

 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

 <span class="c1"># create writer for the generated 3D structures (format specified by file extension)</span>
 <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

 <span class="c1"># export only a single 3D structure (in case of multi-conf. input molecules)</span>
 <span class="n">Chem</span><span class="o">.</span><span class="n">setMultiConfExportParameter</span><span class="p">(</span><span class="n">writer</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

 <span class="c1"># create and initialize an instance of the class ConfGen.StructureGenerator which will</span>
 <span class="c1"># perform the actual 3D structure generation work</span>
 <span class="n">struct_gen</span> <span class="o">=</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">StructureGenerator</span><span class="p">()</span>

 <span class="n">struct_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">timeout</span> <span class="o">=</span> <span class="n">max_time</span> <span class="o">*</span> <span class="mi">1000</span> <span class="c1"># apply the -t argument</span>

 <span class="c1"># dictionary mapping status codes to human readable strings</span>
 <span class="n">status_to_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">UNINITIALIZED</span>                  <span class="p">:</span> <span class="s1">&#39;uninitialized&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TIMEOUT</span>                        <span class="p">:</span> <span class="s1">&#39;max. processing time exceeded&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">ABORTED</span>                        <span class="p">:</span> <span class="s1">&#39;aborted&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_SETUP_FAILED</span>        <span class="p">:</span> <span class="s1">&#39;force field setup failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_MINIMIZATION_FAILED</span> <span class="p">:</span> <span class="s1">&#39;force field structure refinement failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_LIBRARY_NOT_SET</span>       <span class="p">:</span> <span class="s1">&#39;fragment library not available&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_FAILED</span>       <span class="p">:</span> <span class="s1">&#39;fragment conformer generation failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_TIMEOUT</span>      <span class="p">:</span> <span class="s1">&#39;fragment conformer generation timeout&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_ALREADY_PROCESSED</span>     <span class="p">:</span> <span class="s1">&#39;fragment already processed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TORSION_DRIVING_FAILED</span>         <span class="p">:</span> <span class="s1">&#39;torsion driving failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">CONF_GEN_FAILED</span>                <span class="p">:</span> <span class="s1">&#39;conformer generation failed&#39;</span> <span class="p">}</span>

 <span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
 <span class="k">try</span><span class="p">:</span>
     <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
         <span class="c1"># compose a simple molecule identifier</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

         <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
             <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
         <span class="k">else</span><span class="p">:</span>
             <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

         <span class="k">try</span><span class="p">:</span>
             <span class="c1"># generate 3D structure of the read molecule</span>
             <span class="n">status</span> <span class="o">=</span> <span class="n">generate3dConformation</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">struct_gen</span><span class="p">)</span>

             <span class="c1"># check for severe error reported by status code</span>
             <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span><span class="p">:</span>
                 <span class="c1"># enforce the output of 3D coordinates in case of MDL file formats</span>
                 <span class="n">Chem</span><span class="o">.</span><span class="n">setMDLDimensionality</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>

                 <span class="c1"># output the generated 3D structure</span>
                 <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                     <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing 3D structure of molecule </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

         <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
             <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: 3D structure generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

 <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
     <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

 <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="generating-conformation-ensembles">
<h2><span class="section-number">2.2. </span>Generating Conformation Ensembles<a class="headerlink" href="#generating-conformation-ensembles" title="Permalink to this heading"></a></h2>
<p>Sometimes, it’s beneficial to generate multiple conformations for a molecule to capture its flexibility and understand its preferred orientations in
various environments.</p>
<p>The function generate_conformation_ensembles() generates a conformation ensemble for a given molecule using the provided initialized
<a class="reference external" href="../python_api_doc/classCDPL_1_1ConfGen_1_1ConformerGenerator.html">ConfGen.ConformerGenerator instance</a>.
The flags min_rmsd, e_window, and max_confs are used to control the output conformer ensemble size and the conformation quality.
They can be set at the beginning of the example script to the desired values.</p>
<p>Here’s how to generate conformation ensembles for a molecule:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ConfGen</span> <span class="k">as</span> <span class="nn">ConfGen</span>

 <span class="k">def</span> <span class="nf">generateConformationEnsembles</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">,</span> <span class="n">conf_gen</span><span class="p">:</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ConformerGenerator</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">):</span>
     <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Generates a conformation ensemble for the argument molecule using the provided initialized ConfGen.ConformerGenerator instance.</span>

<span class="sd">     Parameters:</span>
<span class="sd">     - mol (Chem.BasicMolecule): Molecule to generate a conformation ensemble for.</span>
<span class="sd">     - conf_gen (ConfGen.ConformerGenerator): Instance of the ConfGen.ConformerGenerator class.</span>

<span class="sd">     Returns:</span>
<span class="sd">     - int: Status code indicating the success of the conformation ensemble generation.</span>
<span class="sd">     - int: Number of generated conformers.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="c1"># prepare the molecule for conformer generation</span>
     <span class="n">ConfGen</span><span class="o">.</span><span class="n">prepareForConformerGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

     <span class="c1"># generate the conformer ensemble</span>
     <span class="n">status</span> <span class="o">=</span> <span class="n">conf_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
     <span class="n">num_confs</span> <span class="o">=</span> <span class="n">conf_gen</span><span class="o">.</span><span class="n">getNumConformers</span><span class="p">()</span>

     <span class="c1"># if sucessful, store the generated conformer ensemble as</span>
     <span class="c1"># per atom 3D coordinates arrays (= the way conformers are represented in CDPKit)</span>
     <span class="k">if</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">SUCCESS</span> <span class="ow">or</span> <span class="n">status</span> <span class="o">==</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TOO_MUCH_SYMMETRY</span><span class="p">:</span>
         <span class="n">conf_gen</span><span class="o">.</span><span class="n">setConformers</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
     <span class="k">else</span><span class="p">:</span>
         <span class="n">num_confs</span> <span class="o">=</span> <span class="mi">0</span>

     <span class="c1"># return status code and the number of generated conformers</span>
     <span class="k">return</span> <span class="p">(</span><span class="n">status</span><span class="p">,</span> <span class="n">num_confs</span><span class="p">)</span>


 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>

 <span class="c1">#Flags</span>
 <span class="n">max_time</span> <span class="o">=</span> <span class="mi">3600</span> <span class="c1"># Max. allowed molecule processing time in seconds (default: 3600 sec)</span>
 <span class="n">min_rmsd</span> <span class="o">=</span> <span class="mf">0.5</span> <span class="c1"># Output conformer RMSD threshold (default: 0.5)</span>
 <span class="n">e_window</span> <span class="o">=</span> <span class="mf">20.0</span> <span class="c1"># Output conformer energy window (default: 20.0)</span>
 <span class="n">max_confs</span> <span class="o">=</span> <span class="mi">100</span> <span class="c1"># Max. output ensemble size (default: 100)</span>


 <span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

 <span class="c1"># create writer for the generated conformer ensembles (format specified by file extension)</span>
 <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

 <span class="c1"># create and initialize an instance of the class ConfGen.ConformerGenerator which</span>
 <span class="c1"># will perform the actual conformer ensemble generation work</span>
 <span class="n">conf_gen</span> <span class="o">=</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ConformerGenerator</span><span class="p">()</span>

 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">timeout</span> <span class="o">=</span> <span class="n">max_time</span> <span class="o">*</span> <span class="mi">1000</span>          <span class="c1"># apply the -t argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">minRMSD</span> <span class="o">=</span> <span class="n">min_rmsd</span>                 <span class="c1"># apply the -r argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">energyWindow</span> <span class="o">=</span> <span class="n">e_window</span>            <span class="c1"># apply the -e argument</span>
 <span class="n">conf_gen</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">maxNumOutputConformers</span> <span class="o">=</span> <span class="n">max_confs</span> <span class="c1"># apply the -n argument</span>

 <span class="c1"># dictionary mapping status codes to human readable strings</span>
 <span class="n">status_to_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">UNINITIALIZED</span>                  <span class="p">:</span> <span class="s1">&#39;uninitialized&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TIMEOUT</span>                        <span class="p">:</span> <span class="s1">&#39;max. processing time exceeded&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">ABORTED</span>                        <span class="p">:</span> <span class="s1">&#39;aborted&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_SETUP_FAILED</span>        <span class="p">:</span> <span class="s1">&#39;force field setup failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FORCEFIELD_MINIMIZATION_FAILED</span> <span class="p">:</span> <span class="s1">&#39;force field structure refinement failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_LIBRARY_NOT_SET</span>       <span class="p">:</span> <span class="s1">&#39;fragment library not available&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_FAILED</span>       <span class="p">:</span> <span class="s1">&#39;fragment conformer generation failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_CONF_GEN_TIMEOUT</span>      <span class="p">:</span> <span class="s1">&#39;fragment conformer generation timeout&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">FRAGMENT_ALREADY_PROCESSED</span>     <span class="p">:</span> <span class="s1">&#39;fragment already processed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">TORSION_DRIVING_FAILED</span>         <span class="p">:</span> <span class="s1">&#39;torsion driving failed&#39;</span><span class="p">,</span>
                   <span class="n">ConfGen</span><span class="o">.</span><span class="n">ReturnCode</span><span class="o">.</span><span class="n">CONF_GEN_FAILED</span>                <span class="p">:</span> <span class="s1">&#39;conformer generation failed&#39;</span> <span class="p">}</span>


 <span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
 <span class="k">try</span><span class="p">:</span>
     <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
         <span class="c1"># compose a simple molecule identifier</span>
         <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

         <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
             <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
         <span class="k">else</span><span class="p">:</span>
             <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

         <span class="k">try</span><span class="p">:</span>
             <span class="c1"># generate conformer ensemble for read molecule</span>
             <span class="n">status</span><span class="p">,</span> <span class="n">num_confs</span> <span class="o">=</span> <span class="n">generateConformationEnsembles</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">conf_gen</span><span class="p">)</span>

             <span class="c1"># output generated ensemble (if available)</span>
             <span class="k">if</span> <span class="n">num_confs</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                 <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                     <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: output of conformer ensemble for molecule </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

         <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
             <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: conformer ensemble generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

 <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
     <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

 <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="pharmacophore-generation-and-processing">
<h1><span class="section-number">3. </span>Pharmacophore Generation and Processing<a class="headerlink" href="#pharmacophore-generation-and-processing" title="Permalink to this heading"></a></h1>
<p>This section is about the processing of pharmacophore models.
The pharmacophore concept is widely used in drug design and cheminformatics to understand the
interactions between ligands and their biological targets.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The module <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1Pharm.html">CDPL.Pharm</a> provides functionality for the generation and processing of pharmacophore models.
The possible pharmacophore input fotats are <cite>PML</cite> or <cite>CDF</cite>.</p>
</div>
<section id="ligand-based-pharmacophore-generation-using-conformations">
<h2><span class="section-number">3.1. </span>Ligand-based Pharmacophore Generation using Conformations<a class="headerlink" href="#ligand-based-pharmacophore-generation-using-conformations" title="Permalink to this heading"></a></h2>
<p>The following example shows how to generate a ligand-based pharmacophore model from a set of molecules
and output it to a PML file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># first call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span>     <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                    <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4_name</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                        <span class="c1"># use the name of the input molecule as pharmacophore name</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>          <span class="c1"># generate the pharmacophore</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">setName</span><span class="p">(</span><span class="n">ph4</span><span class="p">,</span> <span class="n">ph4_name</span><span class="p">)</span>        <span class="c1"># set the pharmacophore name</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create writer for the generated pharmacophores (format specified by file extension)</span>
<span class="n">writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.pml&quot;</span><span class="p">)</span>

<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="c1"># compose a simple molecule identifier</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

            <span class="n">ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>         <span class="c1"># generate pharmacophore</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>   <span class="c1"># output pharmacophore</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing generated pharmacophore </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="d-ligand-based-pharmacophore-generation-for-a-set-of-conformations">
<h2><span class="section-number">3.2. </span>3D Ligand-based Pharmacophore Generation for a set of Conformations<a class="headerlink" href="#d-ligand-based-pharmacophore-generation-for-a-set-of-conformations" title="Permalink to this heading"></a></h2>
<p>The following example shows how to generate a ligand-based pharmacophore model from a set of molecules.
The pharmacophore model is generated using the atom coordinates of the specified conformation of each molecule.
The name of the pharmacophore is set to the name of the corresponding molecule.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">conf_idx</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the molecule using atom coordinates of the specified conformation.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>
<span class="sd">    - conf_idx (int): Index of the conformation to use for the pharmacophore generation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">conf_idx</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>                                    <span class="c1"># for a new molecule</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># first call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span>     <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                    <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4_name</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                        <span class="c1"># use the name of the input molecule as pharmacophore name</span>

    <span class="k">if</span> <span class="n">conf_idx</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>                                   <span class="c1"># if mol is a multi-conf. molecule use atom 3D coordinates of the specified conf.</span>
        <span class="n">ph4_gen</span><span class="o">.</span><span class="n">setAtom3DCoordinatesFunction</span><span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">AtomConformer3DCoordinatesFunctor</span><span class="p">(</span><span class="n">conf_idx</span><span class="p">))</span>
        <span class="n">ph4_name</span> <span class="o">+=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">conf_idx</span><span class="p">)</span>                 <span class="c1"># and append conformer index to the pharmacophore name</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>          <span class="c1"># generate the pharmacophore</span>
    <span class="n">Pharm</span><span class="o">.</span><span class="n">setName</span><span class="p">(</span><span class="n">ph4</span><span class="p">,</span> <span class="n">ph4_name</span><span class="p">)</span>        <span class="c1"># set the pharmacophore name</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># create writer for the generated pharmacophores (format specified by file extension)</span>
<span class="n">writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.pml&quot;</span><span class="p">)</span>

<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="c1"># compose a simple molecule identifier</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="c1"># fallback if name is empty</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="n">num_confs</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getNumConformations</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
        <span class="n">start_conf_idx</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">if</span> <span class="n">num_confs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>      <span class="c1"># test if molecule has conformations</span>
            <span class="n">start_conf_idx</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="c1"># if not, make sure conformer loop body gets executed</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">conf_idx</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start_conf_idx</span><span class="p">,</span> <span class="n">num_confs</span><span class="p">):</span> <span class="c1"># for each conformer</span>
                <span class="n">ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">conf_idx</span><span class="p">)</span>         <span class="c1"># generate pharmacophore</span>

                <span class="k">if</span> <span class="ow">not</span> <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>   <span class="c1"># output pharmacophore</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing generated pharmacophore </span><span class="si">%s</span><span class="s1"> failed&#39;</span> <span class="o">%</span> <span class="n">mol_id</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore generation or output for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="aligne-ligand-based-pharmacophores-for-a-set-of-conformations">
<h2><span class="section-number">3.3. </span>Aligne Ligand-based Pharmacophores for a set of Conformations<a class="headerlink" href="#aligne-ligand-based-pharmacophores-for-a-set-of-conformations" title="Permalink to this heading"></a></h2>
<p>The following example shows how to align a set of ligand-based pharmacophores to a reference pharmacophore.
The function <em>readRefPharmacophore()</em> reads and returns the specified alignment reference pharmacophore.
The function <em>genPharmacophore()</em> generates and returns the pharmacophore of the specified molecule.
The function <em>clearFeatureOrientations()</em> removes feature orientation informations and sets the feature geometry to Pharm.FeatureGeometry.SPHERE.
The flag pos_only, min_pose_rmsd, exhaustive, and num_out_almnts are used to control the alignment process and can
be set at the beginning of the example code.
The flag <em>pos_only</em> is used to control if only the position of features is considered during alignment.
The flag <em>min_pose_rmsd</em> is used to control the minimum required RMSD between two consecutively output molecule alignment poses.
The flag <em>num_out_almnts</em> is used to control the number of top-ranked alignment solutions to output per molecule (default: best alignment solution only).
The flag <em>exhaustive</em> is used to control if an exhaustive alignment search is performed.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">readRefPharmacophore</span><span class="p">(</span><span class="n">filename</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads and returns the specified alignment reference pharmacophore.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - filename (str): Name of the file containing the reference pharmacophore.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Reference pharmacophore.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># create pharmacophore reader instance</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreReader</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

    <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span> <span class="c1"># read reference pharmacophore</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading reference pharmacophore failed&#39;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading reference pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="k">def</span> <span class="nf">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">Pharmacophore</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the pharmacophore of the molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to generate a pharmacophore for.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Pharm.Pharmacophore: Pharmacophore of the argument molecule.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>       <span class="c1"># call utility function preparing the molecule for pharmacophore generation</span>

    <span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">DefaultPharmacophoreGenerator</span><span class="p">()</span>    <span class="c1"># create an instance of the pharmacophore generator default implementation</span>
    <span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>                   <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>

    <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">ph4</span><span class="p">)</span>                         <span class="c1"># generate the pharmacophore</span>

    <span class="k">return</span> <span class="n">ph4</span>

<span class="k">def</span> <span class="nf">clearFeatureOrientations</span><span class="p">(</span><span class="n">ph4</span><span class="p">:</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Removes feature orientation informations and sets the feature geometry to Pharm.FeatureGeometry.SPHERE.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ph4 (Pharm.BasicPharmacophore): Pharmacophore to clear.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">ftr</span> <span class="ow">in</span> <span class="n">ph4</span><span class="p">:</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">clearOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">)</span>
        <span class="n">Pharm</span><span class="o">.</span><span class="n">setGeometry</span><span class="p">(</span><span class="n">ftr</span><span class="p">,</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">SPHERE</span><span class="p">)</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># FLAGS</span>
<span class="n">pos_only</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># True = only position of features is considered during alignment</span>
<span class="n">num_out_almnts</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># Number of top-ranked alignment solutions to output per molecule (default: best alignment solution only)</span>
<span class="n">min_pose_rmsd</span> <span class="o">=</span> <span class="mf">0.0</span> <span class="c1"># Minimum required RMSD between two consecutively output molecule alignment poses</span>
<span class="n">exhaustive</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># Perform an exhaustive alignment search (default: false)</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># read the reference pharmacophore</span>
<span class="n">ref_ph4</span> <span class="o">=</span> <span class="n">readRefPharmacophore</span><span class="p">(</span><span class="s2">&quot;path_to_reference_pharmacophore.pml&quot;</span><span class="p">)</span>

<span class="c1"># create writer for aligned molecules (format specified by file extension)</span>
<span class="n">mol_writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="s2">&quot;path_to_output_file.sdf&quot;</span><span class="p">)</span>

<span class="c1"># create instance of class implementing the pharmacophore alignment algorithm</span>
<span class="n">almnt</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreAlignment</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># True = aligned features have to be within the tolerance spheres of the ref. features</span>

<span class="k">if</span> <span class="n">pos_only</span><span class="p">:</span>                          <span class="c1"># clear feature orientation information</span>
    <span class="n">clearFeatureOrientations</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">)</span>

<span class="n">almnt</span><span class="o">.</span><span class="n">addFeatures</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>               <span class="c1"># set reference features (True = first set = reference)</span>
<span class="n">almnt</span><span class="o">.</span><span class="n">performExhaustiveSearch</span><span class="p">(</span><span class="n">exhaustive</span><span class="p">)</span> <span class="c1"># set minimum number of top. mapped feature pairs</span>

<span class="c1"># create pharmacophore fit score calculator instance</span>
<span class="n">almnt_score</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreFitScore</span><span class="p">()</span>

<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="c1"># compose a simple molecule identifier</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>  <span class="c1"># fallback if name is empty</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">mol_ph4</span> <span class="o">=</span> <span class="n">generatePharmacophore</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>    <span class="c1"># generate input molecule pharmacophore</span>

            <span class="k">if</span> <span class="n">pos_only</span><span class="p">:</span>                  <span class="c1"># clear feature orientation information</span>
                <span class="n">clearFeatureOrientations</span><span class="p">(</span><span class="n">mol_ph4</span><span class="p">)</span>

            <span class="n">almnt</span><span class="o">.</span><span class="n">clearEntities</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>         <span class="c1"># clear features of previously aligned pharmacophore</span>
            <span class="n">almnt</span><span class="o">.</span><span class="n">addFeatures</span><span class="p">(</span><span class="n">mol_ph4</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># specify features of the pharmacophore to align</span>

            <span class="n">almnt_solutions</span> <span class="o">=</span> <span class="p">[]</span>               <span class="c1"># stores the found alignment solutions</span>

            <span class="k">while</span> <span class="n">almnt</span><span class="o">.</span><span class="n">nextAlignment</span><span class="p">():</span>                                     <span class="c1"># iterate over all alignment solutions that can be found</span>
                <span class="n">score</span> <span class="o">=</span> <span class="n">almnt_score</span><span class="p">(</span><span class="n">ref_ph4</span><span class="p">,</span> <span class="n">mol_ph4</span><span class="p">,</span> <span class="n">almnt</span><span class="o">.</span><span class="n">getTransform</span><span class="p">())</span>  <span class="c1"># calculate alignment score</span>
                <span class="n">xform</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Matrix4D</span><span class="p">(</span><span class="n">almnt</span><span class="o">.</span><span class="n">getTransform</span><span class="p">())</span>                  <span class="c1"># make a copy of the alignment transformation (mol. ph4 -&gt; ref. ph4)</span>

                <span class="n">almnt_solutions</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">score</span><span class="p">,</span> <span class="n">xform</span><span class="p">))</span>

            <span class="n">saved_coords</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Vector3DArray</span><span class="p">()</span>      <span class="c1"># create data structure for storing 3D coordinates</span>

            <span class="n">Chem</span><span class="o">.</span><span class="n">get3DCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">saved_coords</span><span class="p">)</span> <span class="c1"># save the original atom coordinates</span>

            <span class="n">struct_data</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">hasStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>           <span class="c1"># get existing structure data if available</span>
                <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>                                    <span class="c1"># otherwise create and set new structure data</span>
                <span class="n">struct_data</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">StringDataBlock</span><span class="p">()</span>

                <span class="n">Chem</span><span class="o">.</span><span class="n">setStructureData</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">strut</span><span class="p">)</span>

            <span class="c1"># add alignment score entry to struct. data</span>
            <span class="n">struct_data</span><span class="o">.</span><span class="n">addEntry</span><span class="p">(</span><span class="s1">&#39;&lt;PharmFitScore&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

            <span class="n">output_cnt</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">last_pose</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># order solutions by desc. alignment score</span>
            <span class="n">almnt_solutions</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">almnt_solutions</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">entry</span><span class="p">:</span> <span class="n">entry</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># output molecule alignment poses until the max. number of best output solutions has been reached</span>
            <span class="k">for</span> <span class="n">solution</span> <span class="ow">in</span> <span class="n">almnt_solutions</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">output_cnt</span> <span class="o">==</span> <span class="n">num_out_almnts</span><span class="p">:</span>
                    <span class="k">break</span>

                <span class="n">curr_pose</span> <span class="o">=</span> <span class="n">Math</span><span class="o">.</span><span class="n">Vector3DArray</span><span class="p">(</span><span class="n">saved_coords</span><span class="p">)</span>

                <span class="n">Math</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span><span class="n">curr_pose</span><span class="p">,</span> <span class="n">solution</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>  <span class="c1"># transform atom coordinates</span>

                <span class="c1"># check whether the current pose is &#39;different enough&#39; from</span>
                <span class="c1"># the last pose to qualify for output</span>
                <span class="k">if</span> <span class="n">min_pose_rmsd</span> <span class="o">&gt;</span> <span class="mf">0.0</span> <span class="ow">and</span> <span class="n">last_pose</span> <span class="ow">and</span> <span class="n">Math</span><span class="o">.</span><span class="n">calcRMSD</span><span class="p">(</span><span class="n">last_pose</span><span class="p">,</span> <span class="n">curr_pose</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">min_pose_rmsd</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="c1"># apply the transformed atom coordinates</span>
                <span class="n">Chem</span><span class="o">.</span><span class="n">set3DCoordinates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">curr_pose</span><span class="p">)</span>

                <span class="c1"># store alignment score in the struct. data entry</span>
                <span class="n">struct_data</span><span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">struct_data</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">setData</span><span class="p">(</span><span class="nb">format</span><span class="p">(</span><span class="n">solution</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;.4f&#39;</span><span class="p">))</span>

                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">mol_writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span> <span class="c1"># output the alignment pose of the molecule</span>
                        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing alignment pose of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe write errors</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing alignment pose of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

                <span class="n">last_pose</span> <span class="o">=</span> <span class="n">curr_pose</span>
                <span class="n">output_cnt</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: pharmacophore alignment of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading input molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">mol_writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="pharmacophore-features-and-feature-types">
<h2><span class="section-number">3.4. </span>Pharmacophore Features and Feature Types<a class="headerlink" href="#pharmacophore-features-and-feature-types" title="Permalink to this heading"></a></h2>
<p>Pharmacophore features are abstract representations of molecular features.
The script below shows how to process the properties of pharmacophore features.
The function print_pharmacophore_properties() outputs all (available) properties of the features stored in the given feature container.
E.g. the feature type, geometry, tolerance, weight, and hydrophobicity.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">print_pharmacophore_properties</span><span class="p">(</span><span class="n">ph4</span><span class="p">:</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainer</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs all (available) properties of the features stored in the given feature container.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ph4 (Pharm.FeatureContainer): Feature container to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ftr_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">UNKNOWN</span>               <span class="p">:</span> <span class="s1">&#39;UNKNOWN&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HYDROPHOBIC</span>           <span class="p">:</span> <span class="s1">&#39;HYDROPHOBIC&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">AROMATIC</span>              <span class="p">:</span> <span class="s1">&#39;AROMATIC&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">NEGATIVE_IONIZABLE</span>    <span class="p">:</span> <span class="s1">&#39;NEGATIVE_IONIZABLE&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">POSITIVE_IONIZABLE</span>    <span class="p">:</span> <span class="s1">&#39;POSITIVE_IONIZABLE&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">H_BOND_DONOR</span>          <span class="p">:</span> <span class="s1">&#39;H_BOND_DONOR&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">H_BOND_ACCEPTOR</span>       <span class="p">:</span> <span class="s1">&#39;H_BOND_ACCEPTOR&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HALOGEN_BOND_DONOR</span>    <span class="p">:</span> <span class="s1">&#39;HALOGEN_BOND_DONOR&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">HALOGEN_BOND_ACCEPTOR</span> <span class="p">:</span> <span class="s1">&#39;HALOGEN_BOND_ACCEPTOR&#39;</span><span class="p">,</span>
                    <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureType</span><span class="o">.</span><span class="n">EXCLUSION_VOLUME</span>      <span class="p">:</span> <span class="s1">&#39;EXCLUSION_VOLUME&#39;</span> <span class="p">}</span>

    <span class="n">geom_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">UNDEF</span>   <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">SPHERE</span>  <span class="p">:</span> <span class="s1">&#39;SPHERE&#39;</span><span class="p">,</span>
                <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">VECTOR</span>  <span class="p">:</span> <span class="s1">&#39;VECTOR&#39;</span><span class="p">,</span>
                <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureGeometry</span><span class="o">.</span><span class="n">PLANE</span>   <span class="p">:</span> <span class="s1">&#39;PLANE&#39;</span> <span class="p">}</span>

    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Composition of pharmacophore </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1">:&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">ph4</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ph4</span><span class="p">)):</span>
        <span class="n">ftr</span> <span class="o">=</span> <span class="n">ph4</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39; - Feature #</span><span class="si">%s</span><span class="s1">:&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">ftr_type_str</span><span class="p">[</span><span class="n">Pharm</span><span class="o">.</span><span class="n">getType</span><span class="p">(</span><span class="n">ftr</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Geometry: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geom_str</span><span class="p">[</span><span class="n">Pharm</span><span class="o">.</span><span class="n">getGeometry</span><span class="p">(</span><span class="n">ftr</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Tolerance: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getTolerance</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Weight: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getWeight</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Optional: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getOptionalFlag</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Disabled: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getDisabledFlag</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Length: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getLength</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Hydrophobicity: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getHydrophobicity</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">has3DCoordinates</span><span class="p">(</span><span class="n">ftr</span><span class="p">):</span>         <span class="c1"># Pharm.Feature derives from Chem.Entity3D - therefore a function from the Chem package is used here!</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Position: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Chem</span><span class="o">.</span><span class="n">get3DCoordinates</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">hasOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;  - Orientation: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">getOrientation</span><span class="p">(</span><span class="n">ftr</span><span class="p">))</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># create reader for input pharmacophores (format specified by file extension)</span>
<span class="n">reader</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">PharmacophoreReader</span><span class="p">(</span><span class="s2">&quot;path_to_input_file.pml&quot;</span><span class="p">)</span>

<span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore interface</span>
<span class="n">ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>

<span class="c1"># read and process pharmacophores one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">ph4</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">print_pharmacophore_properties</span><span class="p">(</span><span class="n">ph4</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading pharmacophore failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="d-structure-based-pharmacophore-generation">
<h2><span class="section-number">3.5. </span>3D Structure-based Pharmacophore Generation<a class="headerlink" href="#d-structure-based-pharmacophore-generation" title="Permalink to this heading"></a></h2>
<p>The following example shows how to generate a 3D structure-based pharmacophore model from a set of molecules.
It also reads and preprocesses the specified receptor structure.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The receptor structure can be in the format <cite>*.mol2, *.pdb, *.mmtf</cite></p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.Pharm</span> <span class="k">as</span> <span class="nn">Pharm</span>

<span class="k">def</span> <span class="nf">processReceptorStructure</span><span class="p">(</span><span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">strip_res_list</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads and preprocesses the specified receptor structure.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - path (str): Path to the receptor structure file.</span>
<span class="sd">    - strip_res_list (bool): Whitespace separated list of PDB three-letter codes specifying residues to remove from the receptor structure (e.g. an existing ligand).</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Chem.Molecule: Receptor structure.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># create reader for receptor structure (format specified by file extension)</span>
    <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="s2">&quot;path_to_receptor_structure_file.pdb&quot;</span><span class="p">)</span><span class="s2">&quot;)</span>

    <span class="n">sup_fmts</span> <span class="o">=</span> <span class="p">[</span> <span class="n">Chem</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MOL2</span><span class="p">,</span>
                <span class="n">Biomol</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">PDB</span><span class="p">,</span>
                <span class="n">Biomol</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MMTF</span> <span class="p">]</span>

    <span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">getDataFormat</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sup_fmts</span><span class="p">:</span>   <span class="c1"># check if the format is supported by this script</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: receptor input file format </span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> not supported&#39;</span> <span class="o">%</span> <span class="n">name_and_ext</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

    <span class="n">rec_mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>    <span class="c1"># create an instance of the default implementation of the</span>
                                    <span class="c1"># Chem.Molecule interface that will store the receptor struct.</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">):</span>  <span class="c1"># read receptor structure</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading receptor structure failed&#39;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading receptor structure failed:</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="c1"># preprocess the receptor structure (removal of residues and</span>
    <span class="c1"># calculation of properties required by the pharm. generation procedure)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># if structure comes from an MOL2 file, convert MOL2 residue data into PDB-style data</span>
        <span class="k">if</span> <span class="n">reader</span><span class="o">.</span><span class="n">getDataFormat</span><span class="p">()</span> <span class="o">==</span> <span class="n">Chem</span><span class="o">.</span><span class="n">DataFormat</span><span class="o">.</span><span class="n">MOL2</span><span class="p">:</span>
            <span class="n">Biomol</span><span class="o">.</span><span class="n">convertMOL2ToPDBResidueInfo</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>

        <span class="n">rem_atoms</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># delete atoms belonging to residues that should be stripped</span>
        <span class="k">if</span> <span class="n">strip_res_list</span><span class="p">:</span>
            <span class="n">atoms_to_rem</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span> <span class="c1"># will store the atoms to delete</span>
            <span class="n">res_to_strip</span> <span class="o">=</span> <span class="p">{</span> <span class="n">tlc</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">tlc</span> <span class="ow">in</span> <span class="n">strip_res_list</span> <span class="p">}</span>

            <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">rec_mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>     <span class="c1"># identify and note atoms belonging to the stripped residues</span>
                <span class="k">if</span> <span class="n">Biomol</span><span class="o">.</span><span class="n">getResidueCode</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">res_to_strip</span><span class="p">:</span>
                    <span class="n">atoms_to_rem</span><span class="o">.</span><span class="n">addAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">atoms_to_rem</span><span class="o">.</span><span class="n">numAtoms</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">rec_mol</span> <span class="o">-=</span> <span class="n">atoms_to_rem</span>    <span class="c1"># delete atoms from the receptor structure</span>
                <span class="n">rem_atoms</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># prepares the receptor structure for pharmacophore generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="n">rem_atoms</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">):</span>                    <span class="c1"># make implicit hydrogens (if any) explicit</span>
            <span class="n">Chem</span><span class="o">.</span><span class="n">calcHydrogen3DCoordinates</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">)</span>               <span class="c1"># calculate 3D coordinates for the added expl. hydrogens</span>
            <span class="n">Biomol</span><span class="o">.</span><span class="n">setHydrogenResidueSequenceInfo</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># set residue information for the added expl. hydrogens</span>

        <span class="n">MolProp</span><span class="o">.</span><span class="n">calcAtomHydrophobicities</span><span class="p">(</span><span class="n">rec_mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>          <span class="c1"># calculate atom hydrophobicity values (needed for hydrophobic</span>
                                                                <span class="c1"># pharm. feature generation)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of receptor structure failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">rec_mol</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="c1"># Flags</span>
<span class="n">strip_res_list</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># Whitespace separated list of PDB three-letter codes specifying residues to remove from the receptor structure (e.g. an existing ligand)</span>
<span class="n">gen_x_vols</span> <span class="o">=</span> <span class="kc">False</span> <span class="c1"># Generate exclusion volume spheres on pharm. feature atoms of interacting residues</span>


<span class="n">lig_mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="n">rec_mol</span> <span class="o">=</span> <span class="n">processReceptorStructure</span><span class="p">(</span><span class="n">path_to_receptor_structure_file</span><span class="p">,</span> <span class="n">strip_res_list</span><span class="p">)</span>          <span class="c1"># read and preprocess the receptor structure</span>
<span class="n">ph4_writer</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">FeatureContainerWriter</span><span class="p">(</span><span class="s2">&quot;path_to_pha_file.pml&quot;</span><span class="p">)</span> <span class="c1"># create writer for the generated pharmacophores (format specified by file extension)</span>

<span class="n">ia_ph4</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">BasicPharmacophore</span><span class="p">()</span>     <span class="c1"># create an instance of the default implementation of the Pharm.Pharmacophore</span>
                                        <span class="c1"># interface that will store the generated pharmacophores</span>

<span class="n">ph4_gen</span> <span class="o">=</span> <span class="n">Pharm</span><span class="o">.</span><span class="n">InteractionPharmacophoreGenerator</span><span class="p">()</span> <span class="c1"># create an instance of the pharmacophore generator</span>

<span class="n">ph4_gen</span><span class="o">.</span><span class="n">addExclusionVolumes</span><span class="p">(</span><span class="n">gen_x_vols</span><span class="p">)</span>        <span class="c1"># specify whether to generate exclusion volume spheres</span>
                                                    <span class="c1"># on pharm. feature atoms of interacting residues</span>
<span class="k">try</span><span class="p">:</span>

    <span class="c1"># read and process ligand molecules one after the other until the end of input has been reached (or a severe error occurs)</span>
    <span class="k">for</span> <span class="n">lig_mol</span> <span class="ow">in</span> <span class="n">lig_reader</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="c1"># compose a simple ligand identifier for messages</span>

        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;#&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>  <span class="c1"># fallback if name is empty or not available</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_id</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\&#39;</span><span class="si">%s</span><span class="se">\&#39;</span><span class="s1"> (#</span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">Pharm</span><span class="o">.</span><span class="n">prepareForPharmacophoreGeneration</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">)</span> <span class="c1"># make ligand ready for pharm. generation</span>

            <span class="n">ph4_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">lig_mol</span><span class="p">,</span> <span class="n">rec_mol</span><span class="p">,</span> <span class="n">ia_ph4</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span> <span class="c1"># generate the pharmacophore (True = extract ligand environment residues on-the-fly)</span>

            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">ph4_writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">ia_ph4</span><span class="p">):</span> <span class="c1"># output pharmacophore</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing interaction pharmacophore of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>               <span class="c1"># handle exception raised in case of severe write errors</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: writing interaction pharmacophore of molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>                   <span class="c1"># handle exception raised in case of severe processing errors</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: interaction pharmacophore generation for molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>                           <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule </span><span class="si">%s</span><span class="s1"> failed: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>

<span class="n">ph4_writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>
<section id="calculation-of-atom-and-bond-properties">
<h1><span class="section-number">4. </span>Calculation of Atom and Bond Properties<a class="headerlink" href="#calculation-of-atom-and-bond-properties" title="Permalink to this heading"></a></h1>
<p>CDPKit provides a convenient way to calculate various properties of atoms and bonds in a molecule.</p>
<section id="calculation-of-atomic-properties">
<h2><span class="section-number">4.1. </span>Calculation of Atomic Properties<a class="headerlink" href="#calculation-of-atomic-properties" title="Permalink to this heading"></a></h2>
<p>Atomic properties provide important chemical information about each atom in a molecule that
is useful for many cheminformatics tasks. The following code snippet shows how to calculate
and output the corresponding properties of each atom of the provided molecular graph. It includes
the calculation of implicit hydrogen counts, atom hybridization states, smallest set of smallest rings,
cycles, aromaticity, and H-bond donor and acceptor atom types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The following example requires the CDPL.Chem and <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1MolProp.html">CDPL.MolProp</a></p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.MolProp</span> <span class="k">as</span> <span class="nn">MolProp</span>

<span class="k">def</span> <span class="nf">outputProperties</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs the corresponding properties of each atom of the provided molecular graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecular graph to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>
    <span class="n">MolProp</span><span class="o">.</span><span class="n">perceiveHBondDonorAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive H-bond donor atom types and set corresponding atom properties</span>
    <span class="n">MolProp</span><span class="o">.</span><span class="n">perceiveHBondAcceptorAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive H-bond acceptor atom types and set corresponding atom properties</span>

    <span class="n">hba_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">UNDEF</span>                   <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">NONE</span>                    <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_H2O</span>                   <span class="p">:</span> <span class="s1">&#39;O_H2O&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_UREA</span>                  <span class="p">:</span> <span class="s1">&#39;O_UREA&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_BARBITURIC_ACID</span>       <span class="p">:</span> <span class="s1">&#39;O_BARBITURIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_URIC_ACID</span>             <span class="p">:</span> <span class="s1">&#39;O_URIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ETHER</span>                 <span class="p">:</span> <span class="s1">&#39;O_ETHER&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_AMIDE</span>                 <span class="p">:</span> <span class="s1">&#39;O_AMIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_N_OXIDE</span>               <span class="p">:</span> <span class="s1">&#39;O_N_OXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ACID</span>                  <span class="p">:</span> <span class="s1">&#39;O_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ESTER</span>                 <span class="p">:</span> <span class="s1">&#39;O_ESTER&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_SULFOXIDE</span>             <span class="p">:</span> <span class="s1">&#39;O_SULFOXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_NITRO</span>                 <span class="p">:</span> <span class="s1">&#39;O_NITRO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_SELEN_OXIDE</span>           <span class="p">:</span> <span class="s1">&#39;O_SELEN_OXIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ALDEHYD</span>               <span class="p">:</span> <span class="s1">&#39;O_ALDEHYD&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_KETONE</span>                <span class="p">:</span> <span class="s1">&#39;O_KETONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">O_ALCOHOL</span>               <span class="p">:</span> <span class="s1">&#39;O_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_NH3</span>                   <span class="p">:</span> <span class="s1">&#39;N_NH3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_DIAMINE</span>               <span class="p">:</span> <span class="s1">&#39;N_DIAMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_MONO_DI_NITRO_ANILINE</span> <span class="p">:</span> <span class="s1">&#39;N_MONO_DI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_TRI_NITRO_ANILINE</span>     <span class="p">:</span> <span class="s1">&#39;N_TRI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_HALOGENO_ANILINE</span>      <span class="p">:</span> <span class="s1">&#39;N_HALOGENO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_ANILINE</span>               <span class="p">:</span> <span class="s1">&#39;N_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_NITRILE</span>               <span class="p">:</span> <span class="s1">&#39;N_NITRILE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZOLE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AZOLE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AMINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AMIDINE</span>               <span class="p">:</span> <span class="s1">&#39;N_AMIDINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZO</span>                   <span class="p">:</span> <span class="s1">&#39;N_AZO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_AZINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_AZINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_DIAZINE</span>               <span class="p">:</span> <span class="s1">&#39;N_DIAZINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">N_IMINE</span>                 <span class="p">:</span> <span class="s1">&#39;N_IMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">S_SULFIDE</span>               <span class="p">:</span> <span class="s1">&#39;S_SULFIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">S_THIOUREA</span>              <span class="p">:</span> <span class="s1">&#39;S_THIOUREA&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">P_MONO_DI_PHOSPHINE</span>     <span class="p">:</span> <span class="s1">&#39;P_MONO_DI_PHOSPHINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondAcceptorAtomType</span><span class="o">.</span><span class="n">P_TRI_PHOSPHINE</span>         <span class="p">:</span> <span class="s1">&#39;P_TRI_PHOSPHINE&#39;</span> <span class="p">}</span>

    <span class="n">hbd_type_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">UNDEF</span>                       <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">NONE</span>                        <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">I_HI</span>                        <span class="p">:</span> <span class="s1">&#39;I_HI&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">BR_HBR</span>                      <span class="p">:</span> <span class="s1">&#39;BR_HBR&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">CL_HCL</span>                      <span class="p">:</span> <span class="s1">&#39;CL_HCL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_HSCN</span>                      <span class="p">:</span> <span class="s1">&#39;S_HSCN&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">F_HF</span>                        <span class="p">:</span> <span class="s1">&#39;F_HF&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">H_H2</span>                        <span class="p">:</span> <span class="s1">&#39;H_H2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">C_HCN</span>                       <span class="p">:</span> <span class="s1">&#39;C_HCN&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">C_ETHINE</span>                    <span class="p">:</span> <span class="s1">&#39;C_ETHINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_HN3</span>                       <span class="p">:</span> <span class="s1">&#39;N_HN3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_NH3</span>                       <span class="p">:</span> <span class="s1">&#39;N_NH3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_NH4</span>                       <span class="p">:</span> <span class="s1">&#39;N_NH4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMINE</span>                     <span class="p">:</span> <span class="s1">&#39;N_AMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMINIUM</span>                   <span class="p">:</span> <span class="s1">&#39;N_AMINIUM&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_ANILINE</span>                   <span class="p">:</span> <span class="s1">&#39;N_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_MONO_DI_NITRO_ANILINE</span>     <span class="p">:</span> <span class="s1">&#39;N_MONO_DI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_TRI_NITRO_ANILINE</span>         <span class="p">:</span> <span class="s1">&#39;N_TRI_NITRO_ANILINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_PYRROLE</span>                   <span class="p">:</span> <span class="s1">&#39;N_PYRROLE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_AMIDE</span>                     <span class="p">:</span> <span class="s1">&#39;N_AMIDE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_IMINE</span>                     <span class="p">:</span> <span class="s1">&#39;N_IMINE&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">N_IMINIUM</span>                   <span class="p">:</span> <span class="s1">&#39;N_IMINIUM&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_H2S</span>                       <span class="p">:</span> <span class="s1">&#39;S_H2S&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_HS</span>                        <span class="p">:</span> <span class="s1">&#39;S_HS&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">S_THIOL</span>                     <span class="p">:</span> <span class="s1">&#39;S_THIOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H3PO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H3PO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2CO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2CO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HCO3</span>                      <span class="p">:</span> <span class="s1">&#39;O_HCO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2O2</span>                      <span class="p">:</span> <span class="s1">&#39;O_H2O2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2O</span>                       <span class="p">:</span> <span class="s1">&#39;O_H2O&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CF3SO3H</span>                   <span class="p">:</span> <span class="s1">&#39;O_CF3SO3H&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HCLO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_HCLO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2SO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2SO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HNO3</span>                      <span class="p">:</span> <span class="s1">&#39;O_HNO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HSO4</span>                      <span class="p">:</span> <span class="s1">&#39;O_HSO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HNO2</span>                      <span class="p">:</span> <span class="s1">&#39;O_HNO2&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_NH2OH</span>                     <span class="p">:</span> <span class="s1">&#39;O_NH2OH&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2PO4</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2PO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H3BO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H3BO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H4SIO4</span>                    <span class="p">:</span> <span class="s1">&#39;O_H4SIO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HPO4</span>                      <span class="p">:</span> <span class="s1">&#39;O_HPO4&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_H2BO3</span>                     <span class="p">:</span> <span class="s1">&#39;O_H2BO3&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HO</span>                        <span class="p">:</span> <span class="s1">&#39;O_HO&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_SULFONIC_ACID</span>             <span class="p">:</span> <span class="s1">&#39;O_SULFONIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_MONO_DI_NITRO_PHENOL</span>      <span class="p">:</span> <span class="s1">&#39;O_MONO_DI_NITRO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_ALCOHOL</span>          <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_ALCOHOL</span>                   <span class="p">:</span> <span class="s1">&#39;O_ALCOHOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_TRI_NITRO_PHENOL</span>          <span class="p">:</span> <span class="s1">&#39;O_TRI_NITRO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_PHENOL</span>           <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_PHENOL</span>                    <span class="p">:</span> <span class="s1">&#39;O_PHENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CARBOXYLIC_ACID</span>           <span class="p">:</span> <span class="s1">&#39;O_CARBOXYLIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_HALOGENO_CARBOXYCLIC_ACID</span> <span class="p">:</span> <span class="s1">&#39;O_HALOGENO_CARBOXYCLIC_ACID&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_ENOL</span>                      <span class="p">:</span> <span class="s1">&#39;O_ENOL&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_OXIME</span>                     <span class="p">:</span> <span class="s1">&#39;O_OXIME&#39;</span><span class="p">,</span>
                    <span class="n">MolProp</span><span class="o">.</span><span class="n">HBondDonorAtomType</span><span class="o">.</span><span class="n">O_CL5_PHENOL</span>                <span class="p">:</span> <span class="s1">&#39;O_CL5_PHENOL&#39;</span> <span class="p">}</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom #</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">molgraph</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is std. hydrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isOrdinaryHydrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is heavy atom: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHeavy</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is unsaturated: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isUnsaturated</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is H-bond acceptor: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHBondAcceptor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">H-bond acceptor type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hba_type_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHBondAcceptorType</span><span class="p">(</span><span class="n">atom</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is H-bond donor: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isHBondDonor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">H-bond donor type: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hbd_type_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHBondDonorType</span><span class="p">(</span><span class="n">atom</span><span class="p">)])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is carbonyl carbon: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isCarbonylLikeAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is amide carbon: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isAmideCenterAtom</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is amide nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isAmideNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">True</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is invertible nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isInvertibleNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Is planar nitrogen: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">isPlanarNitrogen</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Usage: </span><span class="si">%s</span><span class="s1"> &lt;input mol. file&gt;&#39;</span> <span class="o">%</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">outputProperties</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="calculation-of-connectivity-properties">
<h2><span class="section-number">4.2. </span>Calculation of Connectivity Properties<a class="headerlink" href="#calculation-of-connectivity-properties" title="Permalink to this heading"></a></h2>
<p>This code provides a way to calculate various properties of atoms and bonds in a molecule.
The code snippet provided below shows how to calculate and output the corresponding properties of each atom of the provided molecular graph.
It includes the calculation of implicit hydrogen counts, atom hybridization states, smallest set of smallest rings, cycles, aromaticity,
and a variety of other connectivity properties such as the number of connected carbon atoms, heteroatoms, halogens, heavy atoms, chain atoms,
ring atoms, aromatic atoms, incident bonds, and incident single bonds as well as valecny, coordination, and ring sizes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.MolProp</span> <span class="k">as</span> <span class="nn">MolProp</span>

<span class="k">def</span> <span class="nf">outputProperties</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Outputs the corresponding properties of each atom of the provided molecular graph.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecular graph to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">vsepr_geom_str</span> <span class="o">=</span> <span class="p">{</span> <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">UNDEF</span>                  <span class="p">:</span> <span class="s1">&#39;UNDEF&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">NONE</span>                   <span class="p">:</span> <span class="s1">&#39;NONE&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">LINEAR</span>                 <span class="p">:</span> <span class="s1">&#39;LINEAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_PLANAR</span>        <span class="p">:</span> <span class="s1">&#39;TRIGONAL_PLANAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TETRAHEDRAL</span>            <span class="p">:</span> <span class="s1">&#39;TETRAHEDRAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_BIPYRAMIDAL</span>   <span class="p">:</span> <span class="s1">&#39;TRIGONAL_BIPYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">OCTAHEDRAL</span>             <span class="p">:</span> <span class="s1">&#39;OCTAHEDRAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_BIPYRAMIDAL</span> <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_BIPYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_ANTIPRISMATIC</span>   <span class="p">:</span> <span class="s1">&#39;SQUARE_ANTIPRISMATIC&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">BENT</span>                   <span class="p">:</span> <span class="s1">&#39;BENT&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">TRIGONAL_PYRAMIDAL</span>     <span class="p">:</span> <span class="s1">&#39;TRIGONAL_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_PLANAR</span>          <span class="p">:</span> <span class="s1">&#39;SQUARE_PLANAR&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SQUARE_PYRAMIDAL</span>       <span class="p">:</span> <span class="s1">&#39;SQUARE_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">T_SHAPED</span>               <span class="p">:</span> <span class="s1">&#39;T_SHAPED&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">SEESAW</span>                 <span class="p">:</span> <span class="s1">&#39;SEESAW&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_PYRAMIDAL</span>   <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_PYRAMIDAL&#39;</span><span class="p">,</span>
                <span class="n">MolProp</span><span class="o">.</span><span class="n">CoordinationGeometry</span><span class="o">.</span><span class="n">PENTAGONAL_PLANAR</span>      <span class="p">:</span> <span class="s1">&#39;PENTAGONAL_PLANAR&#39;</span> <span class="p">}</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- Atom #</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">molgraph</span><span class="o">.</span><span class="n">getAtomIndex</span><span class="p">(</span><span class="n">atom</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected std. hydrogens (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getOrdinaryHydrogenCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected carbon atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">C</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected heteroatoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">HET</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected halogens: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">Chem</span><span class="o">.</span><span class="n">AtomType</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected heavy atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHeavyAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected chain atoms (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitChainAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected chain atoms (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getChainAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected ring atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRingAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. connected aromatic atoms: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getAromaticAtomCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident single bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident single bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">1</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident double bonds: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">2</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident triple bonds: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">3</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident chain bonds (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getExplicitChainBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident chain bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getChainBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident ring bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRingBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident aromatic bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getAromaticBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident heavy atom bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getHeavyBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Num. incident rotatable bonds (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getRotatableBondCount</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Valence (excl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcExplicitValence</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Valence (incl. impl. H): </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcValence</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Steric number: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">MolProp</span><span class="o">.</span><span class="n">calcStericNumber</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">VSEPR coordination geometry: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">vsepr_geom_str</span><span class="p">[</span><span class="n">MolProp</span><span class="o">.</span><span class="n">getVSEPRCoordinationGeometry</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)])</span>


<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">outputProperties</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="calculation-of-molecule-and-pharmacophore-descriptors">
<h1><span class="section-number">5. </span>Calculation of Molecule and Pharmacophore Descriptors<a class="headerlink" href="#calculation-of-molecule-and-pharmacophore-descriptors" title="Permalink to this heading"></a></h1>
<p>The calculation of molecular descriptors is a fundamental operation in cheminformatics.
It allows for the quantification of various chemical properties of molecules, which can be
used for various cheminformatics tasks, such as:</p>
<ul class="simple">
<li><p><strong>Drug Discovery</strong>: Identifying molecules that contain a particular pharmacophore or active site.</p></li>
<li><p><strong>Chemical Database Querying</strong>: Filtering large chemical databases to retrieve molecules of interest.</p></li>
<li><p><strong>Chemical Analysis</strong>: Identifying the presence of particular functional groups or fragments in molecules.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The CDPKit moduls <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1Descr.html">CDPL.Descr</a> and <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1Util.html">CDPL.Util</a> provide a convenient way to calculate various molecular descriptors.</p>
</div>
<section id="extended-connectivity-fingerprints-ecfps">
<h2><span class="section-number">5.1. </span>Extended Connectivity Fingerprints (ECFPs)<a class="headerlink" href="#extended-connectivity-fingerprints-ecfps" title="Permalink to this heading"></a></h2>
<p>Morgan circular fingerprints, also known as extended connectivity fingerprints (ECFPs),
are a type of structural fingerprints that encode the local chemical environment of each atom in a
molecule. They are widely used in cheminformatics for various tasks, such as similarity searching,
virtual screening, and machine learning.</p>
<p>The following code snippet calculates and outputs the ECFP4 fingerprints of the provided molecules.
Following parameters can be set:</p>
<ul class="simple">
<li><p><strong>num_bits</strong>: The number of bits of the fingerprint (default: 1024)</p></li>
<li><p><strong>radius</strong>: Max. atom environment radius in number of bonds (default: 2)</p></li>
<li><p><strong>inc_hs</strong>: Whether to include explicit hydrogens (by default, the fingerprint is generated for the H-deplete molecular graph)</p></li>
<li><p><strong>inc_config</strong>: Whether to include atom chirality (by default, the fingerprint is generated for the H-deplete molecular graph)</p></li>
</ul>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Descr</span> <span class="k">as</span> <span class="nn">Descr</span>
<span class="kn">import</span> <span class="nn">CDPL.Util</span> <span class="k">as</span> <span class="nn">Util</span>

<span class="k">def</span> <span class="nf">genECFP</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Molecule</span><span class="p">,</span> <span class="n">num_bits</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">radius</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">inc_hs</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">inc_config</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Util</span><span class="o">.</span><span class="n">BitSet</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the binary ECFP for the given molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - mol (Chem.Molecule): Molecule to process.</span>
<span class="sd">    - num_bits (int): Number of bits of the fingerprint.</span>
<span class="sd">    - radius (int): Max. atom environment radius in number of bonds.</span>
<span class="sd">    - inc_hs (bool): Whether to include explicit hydrogens.</span>
<span class="sd">    - inc_config (bool): Whether to include atom chirality.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - Util.BitSet: The generated fingerprint.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># calculate implicit hydrogen counts (if not yet done)</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                      <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                      <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>               <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">ecfp_gen</span> <span class="o">=</span> <span class="n">Descr</span><span class="o">.</span><span class="n">CircularFingerprintGenerator</span><span class="p">()</span>    <span class="c1"># create ECFP generator instance</span>

    <span class="k">if</span> <span class="n">inc_config</span><span class="p">:</span>
        <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">includeChirality</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>                <span class="c1"># allow atom chirality to have an impact on the ECFP generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">calcAtomStereoDescriptors</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>     <span class="c1"># calculate atom stereo descriptors and set corresponding property for all atoms</span>

    <span class="k">if</span> <span class="n">inc_hs</span><span class="p">:</span>
        <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">includeHydrogens</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>                <span class="c1"># include explicit hydrogens in the ECFP generation</span>
        <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                 <span class="c1"># make any implicit hydrogens explicit</span>

    <span class="n">fp</span> <span class="o">=</span> <span class="n">Util</span><span class="o">.</span><span class="n">BitSet</span><span class="p">()</span>                                 <span class="c1"># create fingerprint bitset</span>
    <span class="n">fp</span><span class="o">.</span><span class="n">resize</span><span class="p">(</span><span class="n">num_bits</span><span class="p">)</span>                                <span class="c1"># set desired fingerprint size</span>

    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">setNumIterations</span><span class="p">(</span><span class="n">radius</span><span class="p">)</span>                  <span class="c1"># set num. iterations (=atom. env. radius)</span>
    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">generate</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>                             <span class="c1"># extract chracteristic structural features</span>
    <span class="n">ecfp_gen</span><span class="o">.</span><span class="n">setFeatureBits</span><span class="p">(</span><span class="n">fp</span><span class="p">)</span>                        <span class="c1"># set bits associated with the extracted structural features</span>

    <span class="c1"># if needed, fp could be converted into a numpy single precision float array as follows:</span>
    <span class="c1"># fp = numpy.array(fp, dtype=numpy.float32)</span>

    <span class="k">return</span> <span class="n">fp</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="c1"># set parameters</span>
<span class="n">num_bits</span> <span class="o">=</span> <span class="mi">1024</span>  <span class="c1"># fingerprint size</span>
<span class="n">radius</span> <span class="o">=</span> <span class="mi">4</span>       <span class="c1"># atom environment radius</span>
<span class="n">inc_hs</span> <span class="o">=</span> <span class="kc">True</span>    <span class="c1"># include explicit hydrogens</span>
<span class="n">inc_config</span> <span class="o">=</span> <span class="kc">True</span> <span class="c1"># include atom chirality</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>


<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">fp</span> <span class="o">=</span> <span class="n">genECFP</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">num_bits</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">inc_hs</span><span class="p">,</span> <span class="n">inc_config</span><span class="p">)</span>

            <span class="c1"># do something with the fingerprint</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="n">out_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="fame-atom-environment-fingerprints">
<h2><span class="section-number">5.2. </span>FAME Atom Environment Fingerprints<a class="headerlink" href="#fame-atom-environment-fingerprints" title="Permalink to this heading"></a></h2>
<p>The FAME Atom Environment Fingerprints represent the local environment around atoms in molecules.
It provide a high-resolution view of the molecular landscape, making them especially valuable for tasks like similarity searching, compound clustering,
and structure-activity relationship studies.
Their detailed representation can capture nuances that more generic fingerprints might overlook, offering a more refined perspective on molecular structures.</p>
<p>The following code snippet calculates and outputs the FAME descriptors of the provided molecules.</p>
<p>Following parameters can be set:
- <strong>radius</strong>: Max. atom environment radius in number of bonds (default: 2)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

<span class="k">def</span> <span class="nf">genFAMEDescriptor</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">,</span> <span class="n">radius</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generates the FAME descriptor for the given atom of the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - ctr_atom (Chem.Atom): Atom for which the FAME descriptor is to be calculated.</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecule to process.</span>
<span class="sd">    - radius (int): Max. atom environment radius in number of bonds.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - numpy.array: The generated FAME descriptor.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">env</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">Fragment</span><span class="p">()</span>                                                      <span class="c1"># for storing of extracted environment atoms</span>
    <span class="n">descr</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">Chem</span><span class="o">.</span><span class="n">SybylAtomType</span><span class="o">.</span><span class="n">MAX_TYPE</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">radius</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>

    <span class="n">Chem</span><span class="o">.</span><span class="n">getEnvironment</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="n">radius</span><span class="p">,</span> <span class="n">env</span><span class="p">)</span>                       <span class="c1"># extract environment of center atom reaching</span>
                                                                            <span class="c1"># out up to &#39;radius&#39; bonds</span>
    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">env</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>                                                     <span class="c1"># iterate over extracted environment atoms</span>
        <span class="n">sybyl_type</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getSybylType</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>                                   <span class="c1"># retrieve Sybyl type of environment atom</span>
        <span class="n">top_dist</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getTopologicalDistance</span><span class="p">(</span><span class="n">ctr_atom</span><span class="p">,</span> <span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">)</span>       <span class="c1"># get top. distance between center atom and environment atom</span>
        <span class="n">descr</span><span class="p">[</span><span class="n">top_dist</span> <span class="o">*</span> <span class="p">(</span><span class="n">Chem</span><span class="o">.</span><span class="n">SybylAtomType</span><span class="o">.</span><span class="n">MAX_TYPE</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="n">sybyl_type</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>  <span class="c1"># instead of 1 (= Sybyl type presence) also any other numeric atom</span>
                                                                            <span class="c1"># property could be summed up here</span>
    <span class="k">return</span> <span class="n">descr</span>

<span class="k">def</span> <span class="nf">procMolecule</span><span class="p">(</span><span class="n">molgraph</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraph</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Processes the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">    - molgraph (Chem.MolecularGraph): Molecule to process.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>     <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>    <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                   <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                   <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>            <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSybylAtomTypes</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive Sybyl atom types and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcTopologicalDistanceMatrix</span><span class="p">(</span><span class="n">molgraph</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate topological distance matrix and store as Chem.MolecularGraph property</span>

    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molgraph</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="n">descr</span> <span class="o">=</span> <span class="n">genFAMEDescriptor</span><span class="p">(</span><span class="n">atom</span><span class="p">,</span> <span class="n">molgraph</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>     <span class="c1"># generate atom environment descriptor using a radius of five bonds</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">descr</span><span class="p">)</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>

<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span> <span class="n">of</span> <span class="n">BasicMolecules</span><span class="p">]</span>  <span class="c1"># Example list of BasicMolecules</span>

<span class="c1"># read and process molecules one after the other until the end of input has been reached</span>
<span class="k">try</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">procMolecule</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: processing of molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>

<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> <span class="c1"># handle exception raised in case of severe read errors</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s1">&#39;Error: reading molecule failed: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span>
</pre></div>
</div>
</section>
</section>
<section id="force-field-calculations">
<h1><span class="section-number">6. </span>Force Field Calculations<a class="headerlink" href="#force-field-calculations" title="Permalink to this heading"></a></h1>
<p>Force fields are mathematical models used to predict the molecular mechanics of molecules. They are essential in molecular modeling, helping to estimate the spatial arrangement of atoms in a molecule, their potential energy, and other properties. One of the widely recognized force fields is the Merck Molecular Force Field (MMFF94). It is designed to be applicable to a broad range of molecules, making it versatile for various computational chemistry tasks.</p>
<section id="mmff94-atom-charges-calculation">
<h2><span class="section-number">6.1. </span>MMFF94 Atom Charges Calculation<a class="headerlink" href="#mmff94-atom-charges-calculation" title="Permalink to this heading"></a></h2>
<p>The MMFF94 force field, in particular, provides a method to calculate partial atomic charges, which can be crucial in understanding the electrostatic interactions of a molecule.</p>
<p>The following code snippet calculates and outputs the MMFF94 charges of the atoms for a given list of molecules:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The CDPKit moduls <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1ForceField.html">CDPL.ForceField</a> and <a class="reference external" href="../python_api_doc/namespaceCDPL_1_1Chem.html">CDPL.Chem</a> provide a convenient way to calculate the MMFF94 charges of the atoms for a given molecule.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>
<span class="kn">import</span> <span class="nn">CDPL.ForceField</span> <span class="k">as</span> <span class="nn">ForceField</span>

 <span class="k">def</span> <span class="nf">calc_and_output_charges</span><span class="p">(</span><span class="n">mol</span><span class="p">:</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculates and outputs the MMFF94 charges of the atoms for the provided molecule.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        mol (Chem.BasicMolecule): The molecule for which the MMFF94 charges are to be calculated.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Various preprocessing steps to prepare the molecule</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">calcImplicitHydrogenCounts</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>  <span class="c1"># calculate implicit hydrogen counts and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">makeHydrogenComplete</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>               <span class="c1"># make all implicit hydrogens explicit</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveHybridizationStates</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive atom hybridization states and set corresponding property for all atoms</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">perceiveSSSR</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive smallest set of smallest rings and store as Chem.MolecularGraph property</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setRingFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>                <span class="c1"># perceive cycles and set corresponding atom and bond properties</span>
    <span class="n">Chem</span><span class="o">.</span><span class="n">setAromaticityFlags</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>         <span class="c1"># perceive aromaticity and set corresponding atom and bond properties</span>

    <span class="n">ForceField</span><span class="o">.</span><span class="n">perceiveMMFF94AromaticRings</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>        <span class="c1"># perceive aromatic rings according to the MMFF94 aroamticity model and store data as Chem.MolecularGraph property</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">assignMMFF94AtomTypes</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># perceive MMFF94 atom types (tolerant mode) set corresponding property for all atoms</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">assignMMFF94BondTypeIndices</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span> <span class="c1"># perceive MMFF94 bond types (tolerant mode) set corresponding property for all bonds</span>
    <span class="n">ForceField</span><span class="o">.</span><span class="n">calcMMFF94AtomCharges</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>       <span class="c1"># calculate MMFF94 atom charges (tolerant mode) set corresponding property for all atoms</span>


    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;- MMFF94 partial charges&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">mol</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Atom #</span><span class="si">%s</span><span class="s1">: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">getIndex</span><span class="p">()),</span> <span class="nb">str</span><span class="p">(</span><span class="n">ForceField</span><span class="o">.</span><span class="n">getMMFF94Charge</span><span class="p">(</span><span class="n">atom</span><span class="p">))))</span>

<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">###########################################################################</span>
<span class="c1">################################ Example usage:</span>
<span class="n">mols</span> <span class="o">=</span> <span class="p">[</span><span class="n">list_of_BasicMolecules</span><span class="p">]</span>
<span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
    <span class="n">calc_and_output_charges</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
</pre></div>
</div>
<p>For a deeper understanding of MMFF94 and its applications:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://pubs.acs.org/doi/abs/10.1021/ja9621760">MMFF94 Original Publication</a></p></li>
</ul>
<p>This code provides a way to calculate the MMFF94 charges for each atom in a molecule, which can be essential for various cheminformatics analyses.</p>
</section>
</section>
<section id="substructure-searching-and-matching">
<h1><span class="section-number">7. </span>Substructure Searching and Matching<a class="headerlink" href="#substructure-searching-and-matching" title="Permalink to this heading"></a></h1>
<p>Substructure search is a fundamental operation in cheminformatics. It allows for the identification of molecules that contain a specific structural motif or pattern. This is particularly useful in various applications, such as:</p>
<ul class="simple">
<li><p><strong>Drug Discovery</strong>: Identifying molecules that contain a particular pharmacophore or active site.</p></li>
<li><p><strong>Chemical Database Querying</strong>: Filtering large chemical databases to retrieve molecules of interest.</p></li>
<li><p><strong>Chemical Analysis</strong>: Identifying the presence of particular functional groups or fragments in molecules.</p></li>
</ul>
<section id="filtering-molecules-based-on-a-smarts-pattern">
<h2><span class="section-number">7.1. </span>Filtering Molecules Based on a SMARTS Pattern<a class="headerlink" href="#filtering-molecules-based-on-a-smarts-pattern" title="Permalink to this heading"></a></h2>
<p>The SMARTS notation (SMiles ARbitrary Target Specification) is a language used to describe structural patterns in molecules. It’s an extension of the SMILES notation and allows for more complex and specific pattern descriptions.
CDPKit provides a convenient way to filter molecules that match a specific structural motif described by a SMARTS pattern. This can be useful in various cheminformatics applications, such as database querying, drug discovery, and chemical analysis.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">CDPL.Chem</span> <span class="k">as</span> <span class="nn">Chem</span>

 <span class="k">def</span> <span class="nf">filterMoleculesBySmarts</span><span class="p">(</span><span class="n">input_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">smarts_pattern</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">quiet</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
     <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">     Filters molecules from the input file that match the provided SMARTS pattern and writes them to the output file.</span>

<span class="sd">     Parameters:</span>
<span class="sd">         input_file (str): Path to the input file containing the molecules to be filtered.</span>
<span class="sd">         output_file (str): Path to the output file to which the filtered molecules are written.</span>
<span class="sd">         smarts_pattern (str): SMARTS pattern describing the structural motif to be matched.</span>
<span class="sd">         quiet (bool): If set to True, no progress information is printed to the console.</span>
<span class="sd">     &quot;&quot;&quot;</span>
     <span class="k">try</span><span class="p">:</span>
         <span class="n">sub_srch_ptn</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">parseSMARTS</span><span class="p">(</span><span class="n">smarts_pattern</span><span class="p">)</span>
         <span class="n">Chem</span><span class="o">.</span><span class="n">initSubstructureSearchQuery</span><span class="p">(</span><span class="n">sub_srch_ptn</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error: parsing of SMARTS pattern failed: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
         <span class="k">return</span>

     <span class="n">substr_srch</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">SubstructureSearch</span><span class="p">(</span><span class="n">sub_srch_ptn</span><span class="p">)</span>
     <span class="n">reader</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MoleculeReader</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
     <span class="n">writer</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">MolecularGraphWriter</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>
     <span class="n">mol</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">BasicMolecule</span><span class="p">()</span>
     <span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>

     <span class="k">try</span><span class="p">:</span>
         <span class="k">while</span> <span class="n">reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
             <span class="n">mol_id</span> <span class="o">=</span> <span class="n">Chem</span><span class="o">.</span><span class="n">getName</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">or</span> <span class="sa">f</span><span class="s1">&#39;Molecule_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s1">&#39;</span>

             <span class="n">Chem</span><span class="o">.</span><span class="n">initSubstructureSearchTarget</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>

             <span class="k">if</span> <span class="n">substr_srch</span><span class="o">.</span><span class="n">mappingExists</span><span class="p">(</span><span class="n">mol</span><span class="p">):</span>
                 <span class="k">if</span> <span class="ow">not</span> <span class="n">quiet</span><span class="p">:</span>
                     <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39; -&gt; substructure found, forwarding molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1"> to output file&#39;</span><span class="p">)</span>
                 <span class="n">writer</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
             <span class="k">elif</span> <span class="ow">not</span> <span class="n">quiet</span><span class="p">:</span>
                 <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39; -&gt; substructure not found in molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

             <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

     <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
         <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

     <span class="n">writer</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">###########################################################################</span>
 <span class="c1">################################ Example usage:</span>
 <span class="n">filterMoleculesBySmarts</span><span class="p">(</span><span class="s2">&quot;input.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;output.sdf&quot;</span><span class="p">,</span> <span class="s2">&quot;[#6]1:[#6]:[#6]:[#6]:[#6]:[#6]:1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="CDPL Python Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../cdpl_python_cookbook/index.html" class="btn btn-neutral float-right" title="CDPL Python Cookbook" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Thomas Seidel and Oliver Wieder.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>